// react-base-ui/lib/popup/createPopupHOC.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('./index.less');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var createPopupHOC = function createPopupHOC(WrappedComponent) {
  var _class, _temp;

  return _temp = _class = function (_Component) {
    (0, _inherits3.default)(_class, _Component);

    function _class(props) {
      (0, _classCallCheck3.default)(this, _class);

      var _this = (0, _possibleConstructorReturn3.default)(this, _Component.call(this, props));

      _this.state = {
        show: true
      };
      return _this;
    }

    _class.prototype.dismiss = function dismiss() {
      var _props = this.props,
          _props$popupDismissib = _props.popupDismissible,
          popupDismissible = _props$popupDismissib === undefined ? true : _props$popupDismissib,
          close = _props.close;

      if (popupDismissible && close) {
        close();
      }
    };

    _class.prototype.render = function render() {
      var _this2 = this;

      var _props$popupShowMask = this.props.popupShowMask,
          popupShowMask = _props$popupShowMask === undefined ? true : _props$popupShowMask;


      return _react2.default.createElement(
        'div',
        { className: 'popup' },
        _react2.default.createElement(
          'div',
          {
            className: popupShowMask ? 'popup-desk popup-mask' : 'popup-desk',
            ref: function ref(c) {
              _this2.desk = c;
            },
            onClick: function onClick() {
              return _this2.dismiss();
            }
          },
          _react2.default.createElement(
            'div',
            { className: 'popup-docker' },
            _react2.default.createElement(WrappedComponent, (0, _extends3.default)({}, this.props, { ref: function ref(c) {
                _this2.wrappedComponentInstance = c;
              } }))
          )
        )
      );
    };

    return _class;
  }(_react.Component), _class.propTypes = {
    close: _react.PropTypes.func,
    popupShowMask: _react.PropTypes.bool,
    popupDismissible: _react.PropTypes.bool
  }, _temp;
};

exports.default = createPopupHOC;
module.exports = exports['default'];
