// react-aaui/es/form/PostalCodeInput.js
import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _Object$keys from 'babel-runtime/core-js/object/keys';
import React, { PropTypes } from 'react';

import TextInput from './TextInput';

export var POSTAL_CODE_FIELD_KEY = 'postalCode';

var object = PropTypes.object;

var propTypes = {
  countryConfig: object.isRequired
};

var getPostalCode = function getPostalCode(countryConfig) {
  var regexp = void 0;
  if (countryConfig && countryConfig.addressForm && countryConfig.addressForm.addressFields) {
    var addressFields = countryConfig.addressForm.addressFields;

    var postalCodeFieldKey = _Object$keys(addressFields).filter(function (field) {
      return addressFields[field].addressPart === POSTAL_CODE_FIELD_KEY;
    })[0];
    if (addressFields[postalCodeFieldKey]) {
      regexp = addressFields[postalCodeFieldKey].validationRegex;
    }
  }
  return regexp || '';
};

var PostalCodeInput = function PostalCodeInput(props) {
  var countryConfig = props.countryConfig,
      rest = _objectWithoutProperties(props, ['countryConfig']);

  var regexp = getPostalCode(countryConfig);
  return React.createElement(TextInput, _extends({
    rules: 'regexp:(' + regexp + ')',
    type: 'postalCode'
  }, rest));
};

PostalCodeInput.displayName = 'PostalCodeInput';
PostalCodeInput.propTypes = propTypes;

export default PostalCodeInput;
