// react-aaui/es/form/TextInput.js
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import _extends from 'babel-runtime/helpers/extends';
import React, { PureComponent } from 'react';

import connectForm from './connectForm';
import Input from '../Input';
import { FormFieldAPIPropTypes } from './types';

var propTypes = _extends({}, FormFieldAPIPropTypes);

var TextInput = function (_PureComponent) {
  _inherits(TextInput, _PureComponent);

  function TextInput() {
    var _temp, _this, _ret;

    _classCallCheck(this, TextInput);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _PureComponent.call.apply(_PureComponent, [this].concat(args))), _this), _this.setWrappedComponentInstance = function (input) {
      _this.input = input;
    }, _this.handleBlur = function (e) {
      var _this$props = _this.props,
          onValidate = _this$props.api.onValidate,
          onBlur = _this$props.onBlur;


      onValidate(_this.input.value);

      if (typeof onBlur === 'function') {
        onBlur(e);
      }
    }, _this.handleChange = function (e) {
      var _this$props2 = _this.props,
          setValue = _this$props2.api.setValue,
          onChange = _this$props2.onChange;


      setValue(e.target.value);

      if (typeof onChange === 'function') {
        onChange(e);
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  TextInput.prototype.render = function render() {
    /* eslint-disable no-unused-vars*/
    var _props = this.props,
        setValue = _props.api.setValue,
        fields = _props.fields,
        rest = _objectWithoutProperties(_props, ['api', 'fields']);

    return React.createElement(Input, _extends({
      ref: this.setWrappedComponentInstance,
      type: 'text'
    }, rest, {
      onBlur: this.handleBlur,
      onChange: this.handleChange
    }));
  };

  return TextInput;
}(PureComponent);

TextInput.displayName = 'TextInput';
TextInput.propTypes = propTypes;

export default connectForm()(TextInput);
