// react-aaui/lib/form/connectForm.js
'use strict';

exports.__esModule = true;

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _react = require('react');

var _store = require('./store');

var _store2 = _interopRequireDefault(_store);

var _getDisplayName = require('../shared/getDisplayName');

var _getDisplayName2 = _interopRequireDefault(_getDisplayName);

var _utils = require('../shared/utils');

var _validation = require('./validation');

var _module = require('./module');

var _module2 = _interopRequireDefault(_module);

var _types = require('./types');

var _types2 = require('../shared/types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var aauiFormStore = (0, _keys2.default)(_types.FormStorePropTypes)[0];
var FormFieldConfig = {
  /* eslint-disable no-unused-vars */
  validator: function validator(_ref) {
    var l10n = _ref.l10n;
    return function (validationResult) {
      return validationResult;
    };
  },
  parser: _utils.identity,
  formatter: _utils.identity
};
var connectForm = function connectForm() {
  var formFieldConfig = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return function (WrappedComponent) {
    var connectDisplayName = 'Connect(' + (0, _getDisplayName2.default)(WrappedComponent) + ')';
    var propTypes = (0, _extends3.default)({}, _types.FormFieldPropTypes, _types2.aauiL10nPropTypes);

    var Connect = function (_PureComponent) {
      (0, _inherits3.default)(Connect, _PureComponent);

      function Connect(props, context) {
        (0, _classCallCheck3.default)(this, Connect);

        var _this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call(this, props, context));

        _this.onValidate = function (value) {
          var _this$validate = _this.validate(value),
              errMsg = _this$validate.errMsg;

          _this.setError(errMsg);
        };

        _this.setValue = function (value) {
          _this.store.dispatch((0, _module.setValue)({
            name: _this.props.name,
            value: value
          }));
        };

        _this.setError = function (value) {
          _this.store.dispatch((0, _module.setError)({
            name: _this.props.name,
            value: value
          }));
        };

        _this.setWrappedComponentInstance = function (ref) {
          _this.wrappedComponentInstance = ref;
        };

        _this.getApi = function () {
          return {
            l10n: _this.l10n,
            onValidate: _this.onValidate,
            setValue: _this.setValue,
            setError: _this.setError
          };
        };

        _this.validate = function (value) {
          var _this$props = _this.props,
              name = _this$props.name,
              required = _this$props.required,
              validator = _this$props.validator,
              parser = _this$props.parser,
              _this$props$rules = _this$props.rules,
              rules = _this$props$rules === undefined ? '' : _this$props$rules,
              rest = (0, _objectWithoutProperties3.default)(_this$props, ['name', 'required', 'validator', 'parser', 'rules']);

          var validators = [];

          validators.push(_validation.ruleRunner);

          // Here is for your custom validation function
          if (validator && typeof validator === 'function') {
            validators.push(validator);
          }

          var chain = validators.map(function (validateFunc) {
            return validateFunc((0, _extends3.default)({
              l10n: _this.l10n,
              rules: required ? 'required|' + rules : rules
            }, rest));
          });

          var _reduceReducers = _utils.reduceReducers.apply(undefined, chain)(new _validation.ValidationResult(name, parser(value), null)),
              errMsg = _reduceReducers.errMsg;

          return {
            name: name,
            value: value,
            errMsg: errMsg
          };
        };

        _this.store = props[aauiFormStore] || context[aauiFormStore] || new _store2.default();
        _this.l10n = props.aauiL10n || context.aauiL10n;
        return _this;
      }

      Connect.prototype.componentWillMount = function componentWillMount() {
        if (!this.unregister) {
          this.unregister = this.store.injectReducer(this.props.name, _module2.default);
        }

        if (!this.unregisterValidator && this.store.tryInjectValidator) {
          this.unregisterValidator = this.store.tryInjectValidator(this.props.name, this.validate);
        }
      };

      Connect.prototype.componentWillUnmount = function componentWillUnmount() {
        if (this.unregister) {
          this.unregister(this.props.name);
        }

        this.unregister = null;

        if (this.unregisterValidator) {
          this.unregisterValidator(this.props.name);
        }

        this.unregisterValidator = null;
      };

      Connect.prototype.getWrappedComponentInstance = function getWrappedComponentInstance() {
        return this.wrappedComponentInstance;
      };

      Connect.prototype.render = function render() {
        /* eslint-disable no-unused-vars */
        var _props = this.props,
            required = _props.required,
            validator = _props.validator,
            errMsg = _props.errMsg,
            formatter = _props.formatter,
            parser = _props.parser,
            value = _props.value,
            rules = _props.rules,
            rest = (0, _objectWithoutProperties3.default)(_props, ['required', 'validator', 'errMsg', 'formatter', 'parser', 'value', 'rules']);


        this.renderedElement = (0, _react.createElement)(WrappedComponent, (0, _extends3.default)({
          ref: this.setWrappedComponentInstance,
          api: this.getApi(),
          value: errMsg ? value : formatter(value)
        }, rest));

        return this.renderedElement;
      };

      return Connect;
    }(_react.PureComponent);

    Connect.displayName = connectDisplayName;
    Connect.WrappedComponent = WrappedComponent;
    Connect.propTypes = propTypes;
    Connect.contextTypes = (0, _extends3.default)({}, _types.FormStorePropTypes, _types2.aauiL10nPropTypes);
    Connect.defaultProps = (0, _extends3.default)({}, FormFieldConfig, formFieldConfig);

    return Connect;
  };
};

exports.default = connectForm;
module.exports = exports['default'];
