// react-aaui/lib/form/FormError.js
'use strict';

exports.__esModule = true;
exports.FormError = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _getContext = require('../shared/getContext');

var _getContext2 = _interopRequireDefault(_getContext);

var _types = require('./types');

var _utils = require('./utils');

var _Alert = require('../Alert');

var _Alert2 = _interopRequireDefault(_Alert);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var aauiFormStore = (0, _keys2.default)(_types.FormStorePropTypes)[0];
var bool = _react.PropTypes.bool,
    string = _react.PropTypes.string;

var defaultProps = {
  name: '',
  suppressErrors: true
};
var propTypes = {
  name: string,
  suppressErrors: bool
};

var FormError = exports.FormError = function (_PureComponent) {
  (0, _inherits3.default)(FormError, _PureComponent);

  function FormError(props, context) {
    (0, _classCallCheck3.default)(this, FormError);

    var _this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call(this, props, context));

    _this.handleChange = function () {
      var fields = _this.store.getState();

      var _reduceFields = (0, _utils.reduceFields)(fields),
          errors = _reduceFields.errors,
          isValid = _reduceFields.isValid;

      _this.setState({ errors: errors, isValid: isValid });
    };

    _this.state = (0, _extends3.default)({ errors: {}, isValid: true }, props);
    _this.store = props[aauiFormStore] || context[aauiFormStore];
    return _this;
  }

  FormError.prototype.componentDidMount = function componentDidMount() {
    this.unsubscribe = this.store.subscribe(this.handleChange);
  };

  FormError.prototype.componentWillUnmount = function componentWillUnmount() {
    if (this.unsubscribe) {
      this.unsubscribe();
      this.unsubscribe = null;
    }
  };

  FormError.prototype.render = function render() {
    var _props = this.props,
        name = _props.name,
        suppressErrors = _props.suppressErrors;
    var _state = this.state,
        errors = _state.errors,
        isValid = _state.isValid;

    var filteredErrors = [];
    var fieldNames = (0, _keys2.default)(errors);

    if (name) {
      fieldNames = [name];
    }

    fieldNames.forEach(function (k) {
      var error = errors[k];

      if (error && typeof error === 'string') {
        filteredErrors.push([k, error]);
      }
    });

    return !isValid && _react2.default.createElement(
      _Alert2.default,
      { type: 'danger', noClose: true },
      _react2.default.createElement(
        'strong',
        null,
        'Error'
      ),
      _react2.default.createElement('br', null),
      !suppressErrors && _react2.default.createElement(
        'ul',
        null,
        filteredErrors.map(function (error, index) {
          return _react2.default.createElement(
            'li',
            { key: index },
            error[0],
            ' --- '
          );
        })
      )
    );
  };

  return FormError;
}(_react.PureComponent);

FormError.defaultProps = defaultProps;
FormError.propTypes = propTypes;

exports.default = (0, _getContext2.default)(_types.FormStorePropTypes)(FormError);
