// react-aaui/lib/form/TinyMCEEditor.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _reactTinymce = require('react-tinymce');

var _reactTinymce2 = _interopRequireDefault(_reactTinymce);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _connectForm = require('./connectForm');

var _connectForm2 = _interopRequireDefault(_connectForm);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var string = _react.PropTypes.string,
    func = _react.PropTypes.func,
    shape = _react.PropTypes.shape;

var propTypes = {
  defaultValue: string,
  onChange: func,
  api: shape({
    setValue: func.isRequired,
    onValidate: func.isRequired
  }).isRequired
};

var defaultProps = {
  defaultValue: '',
  config: {
    plugins: 'autolink link image lists print preview textcolor',
    toolbar: 'undo redo | insert | styleselect | fontsizeselect | fontselect | bold italic | forecolor | backcolor | alignleft aligncenter alignright alignjustify | bullist numlist outdent indent | link image',
    textcolor_cols: '5',
    font_formats: 'Arial=arial,helvetica,sans-serif;Courier New=courier new,courier,monospace;AkrutiKndPadmini=Akpdmi-n;ProximaNova=ProximaNova,Arial,sans-serif',
    fontsize_formats: '8pt 10pt 12pt 14pt 18pt 24pt 36pt'
  }
};

var TinyMCEEditor = function (_PureComponent) {
  (0, _inherits3.default)(TinyMCEEditor, _PureComponent);

  function TinyMCEEditor() {
    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, TinyMCEEditor);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call.apply(_PureComponent, [this].concat(args))), _this), _this.onBlur = function (e) {
      var onValidate = _this.props.api.onValidate;

      onValidate(e.target.getContent());
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  TinyMCEEditor.prototype.render = function render() {
    var _props = this.props,
        defaultValue = _props.defaultValue,
        setValue = _props.api.setValue,
        rest = (0, _objectWithoutProperties3.default)(_props, ['defaultValue', 'api']);

    return _react2.default.createElement(_reactTinymce2.default, (0, _extends3.default)({
      content: defaultValue,
      onBlur: this.onBlur,
      onChange: function onChange(e) {
        return setValue(e.target.getContent());
      }
    }, rest));
  };

  return TinyMCEEditor;
}(_react.PureComponent);

TinyMCEEditor.displayName = 'AAUITinyMCEEditor';
TinyMCEEditor.propTypes = propTypes;
TinyMCEEditor.defaultProps = defaultProps;

exports.default = (0, _connectForm2.default)()(TinyMCEEditor);
module.exports = exports['default'];
