// react-aaui/lib/form/types.js
'use strict';

exports.__esModule = true;
exports.FormStorePropTypes = exports.FormFieldPropTypes = exports.FormFieldAPIPropTypes = exports.FormPropTypes = undefined;

var _react = require('react');

var _store = require('./store');

var _store2 = _interopRequireDefault(_store);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var bool = _react.PropTypes.bool,
    string = _react.PropTypes.string,
    object = _react.PropTypes.object,
    func = _react.PropTypes.func,
    shape = _react.PropTypes.shape,
    instanceOf = _react.PropTypes.instanceOf,
    any = _react.PropTypes.any;
var FormPropTypes = exports.FormPropTypes = {
  onSubmit: func,
  onFail: func,
  onChange: func,
  defaultValues: object
};

var FormFieldAPIPropTypes = exports.FormFieldAPIPropTypes = {
  api: shape({
    setValue: func,
    setError: func,
    onValidate: func
  })
};

var FormFieldPropTypes = exports.FormFieldPropTypes = {
  rules: string,
  value: any,
  parser: func,
  formatter: func,
  validator: func,
  errMsg: any,
  fields: object,
  required: bool
};

var FormStorePropTypes = exports.FormStorePropTypes = {
  aauiFormStore: instanceOf(_store2.default)
};
