// react-aaui/lib/form/validation/rules.js
'use strict';

exports.__esModule = true;

var _config = require('../config');

var required = function required(val) {
  return !!val;
};
var email = function email(val) {
  return (/[0-9-._a-z]+@[0-9-.a-z]+\.[a-z]+/.test(val)
  );
};
var max = function max() {
  var val = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  var maxLen = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  return val.length <= maxLen;
};
var min = function min() {
  var val = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  var minLen = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  return val.length >= minLen;
};
var numberic = function numberic(val) {
  return !isNaN(parseFloat(val)) && isFinite(val);
};
var url = function url(val) {
  return (/^((ht|f)tps?):\/\/[\w-]+(\.[\w-]+)+([\w-.,@?^=%&:/~+#]*[\w-@?^=%&/~+#])?$/.test(val)
  );
};
var phone = function phone(val) {
  return _config.PHONE_NUMBER_CONFIG.some(function (phoneNumberPattern) {
    return phoneNumberPattern.test(val);
  });
};

exports.default = {
  required: required,
  email: email,
  max: max,
  min: min,
  numberic: numberic,
  url: url,
  phone: phone
};
module.exports = exports['default'];
