// react-aaui/lib/form/createField.js
'use strict';

exports.__esModule = true;
exports.isRegisteredField = exports.registerFormField = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _getDisplayName = require('../shared/getDisplayName');

var _getDisplayName2 = _interopRequireDefault(_getDisplayName);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var string = _react.PropTypes.string,
    bool = _react.PropTypes.bool,
    func = _react.PropTypes.func,
    oneOfType = _react.PropTypes.oneOfType;

var NON_LAYOUT_FIELD = 'NonLayoutField';
var formFields = [];

var propTypes = {
  errMsg: string,
  label: string,
  required: bool,
  component: oneOfType([func, string]).isRequired
};

var registerFormField = exports.registerFormField = function registerFormField(Field) {
  if (formFields.indexOf(Field) < 0) {
    formFields.push(Field);
  }
};

var isRegisteredField = exports.isRegisteredField = function isRegisteredField(type) {
  return formFields.indexOf(type) >= 0;
};

var createField = function createField(FieldLayout) {
  var Field = function (_PureComponent) {
    (0, _inherits3.default)(Field, _PureComponent);

    function Field() {
      (0, _classCallCheck3.default)(this, Field);
      return (0, _possibleConstructorReturn3.default)(this, _PureComponent.apply(this, arguments));
    }

    Field.prototype.render = function render() {
      var _props = this.props,
          component = _props.component,
          errMsg = _props.errMsg,
          label = _props.label,
          rest = (0, _objectWithoutProperties3.default)(_props, ['component', 'errMsg', 'label']);


      if (FieldLayout) {
        return _react2.default.createElement(
          FieldLayout,
          (0, _extends3.default)({ label: label, errMsg: errMsg, required: this.props.required }, rest),
          (0, _react.createElement)(component, (0, _extends3.default)({ errMsg: errMsg, label: label }, rest))
        );
      }

      return (0, _react.createElement)(component, (0, _extends3.default)({ errMsg: errMsg, label: label }, rest));
    };

    return Field;
  }(_react.PureComponent);

  Field.displayName = 'FormField(' + (0, _getDisplayName2.default)(FieldLayout || NON_LAYOUT_FIELD) + ')';
  Field.propTypes = propTypes;

  registerFormField(Field);

  return Field;
};

exports.default = createField;
