// react-aaui/lib/form/HFormFieldLayout.js
'use strict';

exports.__esModule = true;

var _taggedTemplateLiteralLoose2 = require('babel-runtime/helpers/taggedTemplateLiteralLoose');

var _taggedTemplateLiteralLoose3 = _interopRequireDefault(_taggedTemplateLiteralLoose2);

var _extends3 = require('babel-runtime/helpers/extends');

var _extends4 = _interopRequireDefault(_extends3);

var _templateObject = (0, _taggedTemplateLiteralLoose3.default)(['form__group grid\n        ', ''], ['form__group grid\n        ', '']),
    _templateObject2 = (0, _taggedTemplateLiteralLoose3.default)(['form__label grid-u-2-12 ', ''], ['form__label grid-u-2-12 ', '']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _parseRules = require('./validation/parseRules');

var _parseRules2 = _interopRequireDefault(_parseRules);

var _L10nMessage = require('../shared/L10nMessage');

var _L10nMessage2 = _interopRequireDefault(_L10nMessage);

var _utils = require('../shared/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var string = _react.PropTypes.string,
    bool = _react.PropTypes.bool,
    node = _react.PropTypes.node;


var propTypes = {
  label: string,
  errMsg: string,
  required: bool,
  rules: string,
  children: node
};

/* eslint-disable jsx-a11y/label-has-for */
var HFormFieldLayout = function HFormFieldLayout(_ref) {
  var label = _ref.label,
      errMsg = _ref.errMsg,
      required = _ref.required,
      children = _ref.children,
      _ref$rules = _ref.rules,
      rules = _ref$rules === undefined ? '' : _ref$rules;

  var parsedRules = (0, _parseRules2.default)(rules);
  var l10nMessageValues = parsedRules.reduce(function (r, rule) {
    var _extends2;

    return (0, _extends4.default)({}, r, (_extends2 = {}, _extends2[rule.name] = rule.param, _extends2));
  }, {});

  return _react2.default.createElement(
    'div',
    {
      className: (0, _utils.cls)(_templateObject, errMsg ? 'form__group--error' : '')
    },
    _react2.default.createElement(
      'label',
      { className: (0, _utils.cls)(_templateObject2, required ? 'form__label--require' : '') },
      _react2.default.createElement(_L10nMessage2.default, { id: label })
    ),
    _react2.default.createElement(
      'div',
      { className: 'form__control grid-u-6-12' },
      children
    ),
    errMsg && _react2.default.createElement(
      'div',
      { className: 'form__validate grid-u-4-12' },
      _react2.default.createElement('i', { className: 'icon-close-circle' }),
      _react2.default.createElement(_L10nMessage2.default, { id: errMsg, values: l10nMessageValues })
    )
  );
};

HFormFieldLayout.propTypes = propTypes;

exports.default = HFormFieldLayout;
module.exports = exports['default'];
