// react-aaui/es/Infobar.js
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { PureComponent, PropTypes } from 'react';
import 'active.css/less/components/alert.less';

var bool = PropTypes.bool,
    string = PropTypes.string,
    object = PropTypes.object,
    func = PropTypes.func,
    any = PropTypes.any;


var propTypes = {
  className: string,
  style: object,
  children: any,
  noClose: bool,
  onClose: func
};

var defaultProps = {
  noClose: false
};

var Infobar = function (_PureComponent) {
  _inherits(Infobar, _PureComponent);

  function Infobar() {
    _classCallCheck(this, Infobar);

    return _possibleConstructorReturn(this, _PureComponent.apply(this, arguments));
  }

  Infobar.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        style = _props.style,
        children = _props.children,
        noClose = _props.noClose,
        onClose = _props.onClose;


    return React.createElement(
      'div',
      {
        className: 'infobar ' + (className || ''),
        style: style
      },
      children,
      noClose || React.createElement(
        'button',
        { type: 'button', className: 'close', onClick: onClose },
        '\xD7'
      )
    );
  };

  return Infobar;
}(PureComponent);

export default Infobar;


Infobar.propTypes = propTypes;
Infobar.defaultProps = defaultProps;
Infobar.displayName = 'AAUIInfobar';
