// react-aaui/es/Label.js
import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { PureComponent, PropTypes } from 'react';
import 'active.css/less/components/label.less';

var string = PropTypes.string,
    object = PropTypes.object,
    oneOf = PropTypes.oneOf,
    node = PropTypes.node;

var propTypes = {
  className: string,
  style: object,
  type: oneOf(['success', 'warning', 'danger', 'info']).isRequired,
  children: node
};
var defaultProps = {
  type: 'info'
};

var Label = function (_PureComponent) {
  _inherits(Label, _PureComponent);

  function Label() {
    _classCallCheck(this, Label);

    return _possibleConstructorReturn(this, _PureComponent.apply(this, arguments));
  }

  Label.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        style = _props.style,
        type = _props.type,
        children = _props.children,
        rest = _objectWithoutProperties(_props, ['className', 'style', 'type', 'children']);

    return React.createElement(
      'span',
      _extends({}, rest, {
        className: 'label label-' + type + ' ' + (className || ''),
        style: style
      }),
      children
    );
  };

  return Label;
}(PureComponent);

export default Label;


Label.displayName = 'AAUILabel';
Label.propTypes = propTypes;
Label.defaultProps = defaultProps;
