// react-aaui/es/Pagination.js
import _extends from 'babel-runtime/helpers/extends';
import _taggedTemplateLiteralLoose from 'babel-runtime/helpers/taggedTemplateLiteralLoose';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _templateObject = _taggedTemplateLiteralLoose(['', '\n                       ', '\n                       ', ''], ['', '\n                       ', '\n                       ', '']),
    _templateObject2 = _taggedTemplateLiteralLoose(['', ''], ['', '']);

import React, { PureComponent, PropTypes } from 'react';
import 'active.css/less/components/pagination.less';

import { cls } from './shared/utils';

function pageData(type, theme, total, current, around, url, roundText, startText, endText) {
  function pageLink(page) {
    if (typeof url === 'function') {
      return url(page);
    }

    var pageLinkVar = typeof url === 'string' ? url : '';
    if (pageLinkVar.indexOf('?') > -1) {
      pageLinkVar += '&page';
    } else {
      pageLinkVar += '?page';
    }
    return pageLinkVar + '=' + page;
  }

  var start = 1;
  var end = total;
  var aroundStart = Math.max(current - around, start + 1);
  var aroundEnd = Math.min(current + around, end - 1);
  var data = [];

  switch (type) {
    case 'pager':
      if (current > start) {
        data.push({ text: startText, href: pageLink(current - 1) });
      }
      if (current < end) {
        data.push({ text: endText, href: pageLink(current + 1) });
      }
      break;
    case 'pagination':
      if (current > start) {
        data.push({ text: startText, href: pageLink(current - 1) });
      }
      /* eslint prefer-template: 0 */
      data.push({ text: '' + start, href: pageLink(start) });
      if (aroundStart - start > 1) {
        data.push({ text: roundText });
      }
      for (var i = aroundStart; i < current; i += 1) {
        data.push({ text: '' + i, href: pageLink(i) });
      }
      for (var _i = current !== start ? current : current + 1; _i <= aroundEnd; _i += 1) {
        data.push({ text: '' + _i, href: pageLink(_i) });
      }
      if (end - aroundEnd > 1) {
        data.push({ text: roundText });
      }
      data.push({ text: '' + end, href: pageLink(end) });
      if (current < end) {
        data.push({ text: endText, href: pageLink(current + 1) });
      }
      break;
    default:
      break;
  }

  return data;
}

var number = PropTypes.number,
    string = PropTypes.string,
    object = PropTypes.object,
    oneOfType = PropTypes.oneOfType;


var propTypes = {
  className: string,
  type: string,
  theme: string,
  total: number,
  current: number,
  around: number,
  endText: oneOfType([object, string]),
  linkElement: string,
  startText: oneOfType([object, string]),
  roundText: oneOfType([object, string]),
  url: string
};

var defaultProps = {
  type: 'pagination',
  url: '',
  around: 3,
  linkElement: 'a',
  roundText: '...',
  startText: React.createElement('span', { className: 'icon-uniF053' }),
  endText: React.createElement('span', { className: 'icon-uniF054' })
};

var Pagination = function (_PureComponent) {
  _inherits(Pagination, _PureComponent);

  function Pagination() {
    var _temp, _this, _ret;

    _classCallCheck(this, Pagination);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _PureComponent.call.apply(_PureComponent, [this].concat(args))), _this), _this.render = function () {
      var _this$props = _this.props,
          type = _this$props.type,
          theme = _this$props.theme,
          total = _this$props.total,
          current = _this$props.current,
          around = _this$props.around,
          url = _this$props.url,
          roundText = _this$props.roundText,
          startText = _this$props.startText,
          endText = _this$props.endText,
          A = _this$props.linkElement,
          className = _this$props.className,
          attrs = _objectWithoutProperties(_this$props, ['type', 'theme', 'total', 'current', 'around', 'url', 'roundText', 'startText', 'endText', 'linkElement', 'className']);

      return React.createElement(
        'ul',
        _extends({
          className: cls(_templateObject, type, type === 'pagination' && theme ? 'pagination--' + theme : '', className || '')
        }, attrs),
        pageData(type, theme, total, current, around, url, roundText, startText, endText).map(function (_ref, i) {
          var text = _ref.text,
              href = _ref.href;

          if (href && +text !== current) {
            return React.createElement(
              'li',
              { key: i, className: cls(_templateObject2, +text === current ? 'active' : '') },
              React.createElement(
                A,
                { key: i, href: href },
                text
              )
            );
          } else {
            /* eslint no-else-return: 0 */
            return React.createElement(
              'li',
              {
                key: i,
                className: cls(_templateObject2, +text === current ? 'active' : '')
              },
              React.createElement(
                'span',
                { key: i },
                text
              )
            );
          }
        })
      );
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  return Pagination;
}(PureComponent);

export default Pagination;


Pagination.displayName = 'AAUIPagination';
Pagination.propTypes = propTypes;
Pagination.defaultProps = defaultProps;
