// react-aaui/es/TabsContainer.js
import _taggedTemplateLiteralLoose from 'babel-runtime/helpers/taggedTemplateLiteralLoose';

var _templateObject = _taggedTemplateLiteralLoose(['', ' ', ''], ['', ' ', '']);

import React, { PropTypes } from 'react';

import { cls } from './shared/utils';
import { tabsAPIShape } from './shared/types';

var string = PropTypes.string,
    object = PropTypes.object,
    any = PropTypes.any;


var propTypes = {
  className: string,
  style: object,
  name: string,
  children: any,
  aauiTabsAPI: tabsAPIShape
};

var TabsContainer = function TabsContainer(_ref) {
  var className = _ref.className,
      style = _ref.style,
      name = _ref.name,
      children = _ref.children,
      getSelected = _ref.aauiTabsAPI.getSelected;
  return React.createElement(
    'div',
    {
      className: cls(_templateObject, name !== getSelected() ? 'u-hidden' : '', className || ''),
      style: style
    },
    children
  );
};

TabsContainer.displayName = 'AAUITabsContainer';
TabsContainer.propTypes = propTypes;

export default TabsContainer;
