// react-aaui/es/TabsTitle.js
import React, { PropTypes } from 'react';

import { tabsAPIShape } from './shared/types';

var string = PropTypes.string,
    object = PropTypes.object,
    any = PropTypes.any;


var propTypes = {
  className: string,
  style: object,
  name: string,
  children: any,
  aauiTabsAPI: tabsAPIShape
};

var TabsTitle = function TabsTitle(_ref) {
  var name = _ref.name,
      style = _ref.style,
      className = _ref.className,
      children = _ref.children,
      _ref$aauiTabsAPI = _ref.aauiTabsAPI,
      select = _ref$aauiTabsAPI.select,
      getSelected = _ref$aauiTabsAPI.getSelected;
  return React.createElement(
    'li',
    null,
    React.createElement(
      'a',
      {
        className: (name === getSelected() ? 'active' : '') + ' ' + (className || ''),
        style: style,
        onClick: function onClick() {
          return select(name);
        }
      },
      children
    )
  );
};

TabsTitle.displayName = 'AAUITabsTitle';
TabsTitle.propTypes = propTypes;

export default TabsTitle;
