// react-aaui/es/Tag.js
import _extends from 'babel-runtime/helpers/extends';
import _taggedTemplateLiteralLoose from 'babel-runtime/helpers/taggedTemplateLiteralLoose';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _templateObject = _taggedTemplateLiteralLoose(['tag-input\n        ', ''], ['tag-input\n        ', '']);

import React, { PureComponent, PropTypes } from 'react';
import 'active.css/less/components/tag.less';

import { cls } from './shared/utils';
import KEY_CODES from './shared/keyCodes';

var string = PropTypes.string,
    func = PropTypes.func,
    bool = PropTypes.bool;

var propTypes = {
  className: string,
  value: string,
  onChange: func,
  onCancel: func,
  onClose: func,
  editMode: bool,
  isNew: bool
};
var defaultProps = {
  editMode: false
};

var Tag = function (_PureComponent) {
  _inherits(Tag, _PureComponent);

  function Tag(props) {
    _classCallCheck(this, Tag);

    var _this = _possibleConstructorReturn(this, _PureComponent.call(this, props));

    _this.handleKeyDown = function (e) {
      var target = e.target;

      switch (e.keyCode) {
        case KEY_CODES.ESCAPE:
          e.preventDefault();
          _this.setState({
            isInputCancel: true,
            isTabEnter: false
          }, function () {
            target.blur();
          });
          break;
        case KEY_CODES.TAB:
        case KEY_CODES.ENTER:
          e.preventDefault();

          _this.setState({
            isTabEnter: true
          }, function () {
            if (target) {
              target.blur();
            }
          });
          break;
        default:
          break;
      }
    };

    _this.handleBlur = function (e) {
      var value = e.target.value;
      _this.setState({
        isInput: false
      }, function () {
        if (!_this.state.isInputCancel) {
          var onChange = _this.props.onChange;
          if (onChange) onChange(value, _this.state.isTabEnter);
        } else {
          var onCancel = _this.props.onCancel;
          if (onCancel) onCancel();
        }
      });
    };

    _this.toInput = function (e) {
      e && e.stopPropagation && e.stopPropagation();

      _this.setState({
        isInput: true,
        isInputCancel: false
      }, function () {
        var input = _this.input;

        if (input) {
          input.focus();
          input.select();
        }
      });
    };

    _this.state = {
      isInput: false,
      isTabEnter: false
    };
    return _this;
  }

  Tag.prototype.render = function render() {
    var _this2 = this;

    /* eslint-disable no-unused-vars */
    var _props = this.props,
        onCancel = _props.onCancel,
        className = _props.className,
        editMode = _props.editMode,
        onClose = _props.onClose,
        onChange = _props.onChange,
        value = _props.value,
        isNew = _props.isNew,
        rest = _objectWithoutProperties(_props, ['onCancel', 'className', 'editMode', 'onClose', 'onChange', 'value', 'isNew']);

    var isInput = this.state.isInput;


    if (!editMode) {
      return React.createElement(
        'span',
        { className: 'badge' },
        value
      );
    }

    return React.createElement(
      'span',
      _extends({
        className: cls(_templateObject, className || '')
      }, rest),
      isInput ? React.createElement('input', {
        className: 'input',
        ref: function ref(input) {
          _this2.input = input;
        },
        placeholder: isNew ? value : undefined,
        defaultValue: isNew ? undefined : value,
        onKeyDown: this.handleKeyDown,
        onBlur: this.handleBlur
      }) : React.createElement(
        'span',
        { className: 'badge', onClick: onClose },
        React.createElement(
          'span',
          { onClick: this.toInput },
          value
        )
      )
    );
  };

  return Tag;
}(PureComponent);

export default Tag;


Tag.displayName = 'AAUITag';
Tag.propTypes = propTypes;
Tag.defaultProps = defaultProps;
