// react-aaui/es/form/HFormFieldLayout.js
import _taggedTemplateLiteralLoose from 'babel-runtime/helpers/taggedTemplateLiteralLoose';
import _extends from 'babel-runtime/helpers/extends';

var _templateObject = _taggedTemplateLiteralLoose(['form__group grid\n        ', ''], ['form__group grid\n        ', '']),
    _templateObject2 = _taggedTemplateLiteralLoose(['form__label grid-u-2-12 ', ''], ['form__label grid-u-2-12 ', '']);

import React, { PropTypes } from 'react';

import parseRules from './validation/parseRules';
import L10nMessage from '../shared/L10nMessage';
import { cls } from '../shared/utils';

var string = PropTypes.string,
    bool = PropTypes.bool,
    node = PropTypes.node;


var propTypes = {
  label: string,
  errMsg: string,
  required: bool,
  rules: string,
  children: node
};

/* eslint-disable jsx-a11y/label-has-for */
var HFormFieldLayout = function HFormFieldLayout(_ref) {
  var label = _ref.label,
      errMsg = _ref.errMsg,
      required = _ref.required,
      children = _ref.children,
      _ref$rules = _ref.rules,
      rules = _ref$rules === undefined ? '' : _ref$rules;

  var parsedRules = parseRules(rules);
  var l10nMessageValues = parsedRules.reduce(function (r, rule) {
    var _extends2;

    return _extends({}, r, (_extends2 = {}, _extends2[rule.name] = rule.param, _extends2));
  }, {});

  return React.createElement(
    'div',
    {
      className: cls(_templateObject, errMsg ? 'form__group--error' : '')
    },
    React.createElement(
      'label',
      { className: cls(_templateObject2, required ? 'form__label--require' : '') },
      React.createElement(L10nMessage, { id: label })
    ),
    React.createElement(
      'div',
      { className: 'form__control grid-u-6-12' },
      children
    ),
    errMsg && React.createElement(
      'div',
      { className: 'form__validate grid-u-4-12' },
      React.createElement('i', { className: 'icon-close-circle' }),
      React.createElement(L10nMessage, { id: errMsg, values: l10nMessageValues })
    )
  );
};

HFormFieldLayout.propTypes = propTypes;

export default HFormFieldLayout;
module.exports = exports['default'];
