// react-base-ui/src/popup/createPopupHOC.jsx
import React, { Component, PropTypes } from 'react';

import './index.less';

const createPopupHOC = WrappedComponent => class extends Component {

  static propTypes = {
    close: PropTypes.func,
    popupShowMask: PropTypes.bool,
    popupDismissible: PropTypes.bool
  }

  constructor(props) {
    super(props);

    this.state = {
      show: true
    };
  }

  dismiss() {
    const { popupDismissible = true, close } = this.props;
    if (popupDismissible && close) {
      close();
    }
  }

  render() {
    const { popupShowMask = true } = this.props;

    return (
      <div className="popup">
        <div
          className={popupShowMask ? 'popup-desk popup-mask' : 'popup-desk'}
          ref={(c) => { this.desk = c; }}
          onClick={() => this.dismiss()}
        >
          <div className="popup-docker">
            <WrappedComponent {...this.props} ref={(c) => { this.wrappedComponentInstance = c; }} />
          </div>
        </div>
      </div>
    );
  }
};

export default createPopupHOC;
