// react-base-ui/lib/messages/index.js
'use strict';

exports.__esModule = true;
exports.configMessageService = exports.cleanMessagesByType = exports.cleanMessages = exports.showError = exports.showWarning = exports.showInfo = exports.showSuccess = exports.MessageBoard = exports.MessageTypeEnum = exports.MessageGroup = undefined;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _emitter = require('./emitter');

var _emitter2 = _interopRequireDefault(_emitter);

var _MessageGroup = require('./MessageGroup');

var _MessageGroup2 = _interopRequireDefault(_MessageGroup);

var _MessageTypeEnum = require('./consts/MessageTypeEnum');

var MessageTypeEnum = _interopRequireWildcard(_MessageTypeEnum);

var _MessageBoard = require('./components/MessageBoard');

var _MessageBoard2 = _interopRequireDefault(_MessageBoard);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.MessageGroup = _MessageGroup2.default;
exports.MessageTypeEnum = MessageTypeEnum;
exports.MessageBoard = _MessageBoard2.default;


var defaultOptions = {
  noDuplicated: true,
  autoClean: true,
  dismissable: false,
  topOnly: true
};

var createService = function createService(type) {
  return function (messages) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    options = (0, _extends3.default)({}, defaultOptions, { options: options });
    var msgGroup = messages instanceof _MessageGroup2.default ? messages : new _MessageGroup2.default(type, messages);
    msgGroup.type = type;
    _emitter2.default.emit('add/notify', msgGroup, options);

    return msgGroup.id;
  };
};

var showSuccess = exports.showSuccess = createService(MessageTypeEnum.SUCCESS);
var showInfo = exports.showInfo = createService(MessageTypeEnum.INFO);
var showWarning = exports.showWarning = createService(MessageTypeEnum.WARNING);
var showError = exports.showError = createService(MessageTypeEnum.ERROR);
var cleanMessages = exports.cleanMessages = function cleanMessages() {
  return _emitter2.default.emit('clean/notify', '');
};
var cleanMessagesByType = exports.cleanMessagesByType = function cleanMessagesByType(type) {
  return _emitter2.default.emit('clean/notify', type);
};
var configMessageService = exports.configMessageService = function configMessageService(options) {
  defaultOptions = (0, _assign2.default)(defaultOptions, options);
};
