// react-base-ui/lib/messages/components/Messages/index.js
'use strict';

exports.__esModule = true;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _unescape = require('lodash/unescape');

var _unescape2 = _interopRequireDefault(_unescape);

var _isString = require('lodash/isString');

var _isString2 = _interopRequireDefault(_isString);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Alert = require('react-aaui/lib/Alert');

var _Alert2 = _interopRequireDefault(_Alert);

var _MessageTypeEnum = require('../../consts/MessageTypeEnum');

var MessageTypeEnum = _interopRequireWildcard(_MessageTypeEnum);

require('./index.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Messages = function (_PureComponent) {
  (0, _inherits3.default)(Messages, _PureComponent);

  function Messages() {
    (0, _classCallCheck3.default)(this, Messages);
    return (0, _possibleConstructorReturn3.default)(this, _PureComponent.apply(this, arguments));
  }

  Messages.prototype.render = function render() {
    var _props = this.props,
        msgGroup = _props.msgGroup,
        _onClose = _props.onClose;
    var _msgGroup$messages = msgGroup.messages,
        messages = _msgGroup$messages === undefined ? [] : _msgGroup$messages,
        type = msgGroup.type,
        id = msgGroup.id,
        title = msgGroup.title,
        _msgGroup$dismissable = msgGroup.dismissable,
        dismissable = _msgGroup$dismissable === undefined ? false : _msgGroup$dismissable;

    var alertType = type === MessageTypeEnum.ERROR ? 'danger' : type;
    var contentClass = messages.length > 1 ? 'message-content' : 'message-content single-line';

    return _react2.default.createElement(
      _Alert2.default,
      {
        key: id,
        type: alertType,
        noClose: !dismissable,
        onClose: function onClose() {
          return _onClose(id);
        },
        className: 'message-box'
      },
      _react2.default.createElement(
        'div',
        { className: contentClass },
        title ? _react2.default.createElement(
          'span',
          { className: 'title' },
          title
        ) : undefined,
        _react2.default.createElement(
          'ul',
          null,
          messages.map(function (line, i) {
            if ((0, _isString2.default)(line)) {
              return _react2.default.createElement(
                'li',
                { key: i },
                (0, _unescape2.default)(line)
              );
            }

            return _react2.default.createElement(
              'li',
              { key: i },
              line
            );
          })
        )
      )
    );
  };

  return Messages;
}(_react.PureComponent); //  eslint-disable-line no-unused-vars


Messages.displayName = 'Messages';
Messages.propTypes = {
  msgGroup: _react.PropTypes.shape({}),
  onClose: _react.PropTypes.func
};
Messages.defaultProps = {};
exports.default = Messages;
module.exports = exports['default'];
