// react-aaui/es/shared/L10nProvider.js
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import _extends from 'babel-runtime/helpers/extends';
import { PureComponent, PropTypes, Children } from 'react';

import L10n from './L10n';
import defaultValidationMessages from '../form/i18n/ValidationMsg.json';
import { aauiL10nPropTypes } from './types';

var string = PropTypes.string,
    object = PropTypes.object,
    node = PropTypes.node;

var propTypes = {
  locale: string,
  messages: object,
  children: node
};
var defaultProps = {
  locale: 'en_US'
};
var childContextTypes = _extends({}, aauiL10nPropTypes);

var L10nProvider = function (_PureComponent) {
  _inherits(L10nProvider, _PureComponent);

  function L10nProvider(props, context) {
    _classCallCheck(this, L10nProvider);

    var _this = _possibleConstructorReturn(this, _PureComponent.call(this, props, context));

    _this.l10n = new L10n(props, context);
    return _this;
  }

  L10nProvider.prototype.getChildContext = function getChildContext() {
    return {
      aauiL10n: this.l10n
    };
  };

  L10nProvider.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    if (this.props.locale !== nextProps.locale || this.props.messages !== nextProps.messages) {
      this.l10n.locale = nextProps.locale;
      this.l10n.messages = _extends({}, this.l10n.messages, defaultValidationMessages[nextProps.locale], nextProps.messages);
    }
  };

  L10nProvider.prototype.render = function render() {
    return Children.only(this.props.children);
  };

  return L10nProvider;
}(PureComponent);

export default L10nProvider;


L10nProvider.propTypes = propTypes;
L10nProvider.defaultProps = defaultProps;
L10nProvider.childContextTypes = childContextTypes;
module.exports = exports['default'];
