// react-base-ui/lib/confirmation/components/Dialog/index.js
'use strict';

exports.__esModule = true;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Button = require('react-aaui/lib/Button');

var _Button2 = _interopRequireDefault(_Button);

var _Modal = require('react-aaui/lib/Modal');

var _Modal2 = _interopRequireDefault(_Modal);

require('active.css/less/components/modal.less');

require('./index.less');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Dialog = function (_React$PureComponent) {
  (0, _inherits3.default)(Dialog, _React$PureComponent);

  function Dialog() {
    (0, _classCallCheck3.default)(this, Dialog);
    return (0, _possibleConstructorReturn3.default)(this, _React$PureComponent.apply(this, arguments));
  }

  Dialog.prototype.render = function render() {
    var _props = this.props,
        _props$title = _props.title,
        title = _props$title === undefined ? 'Confirm' : _props$title,
        _props$message = _props.message,
        message = _props$message === undefined ? '' : _props$message,
        show = _props.show,
        dismiss = _props.dismiss,
        cancel = _props.cancel,
        proceed = _props.proceed,
        _props$showCancel = _props.showCancel,
        showCancel = _props$showCancel === undefined ? false : _props$showCancel,
        _props$cancelText = _props.cancelText,
        cancelText = _props$cancelText === undefined ? 'Cancel' : _props$cancelText,
        _props$confirmText = _props.confirmText,
        confirmText = _props$confirmText === undefined ? 'OK' : _props$confirmText;


    return _react2.default.createElement(
      _Modal2.default,
      { shown: show, title: title, className: 'dialog', onClose: dismiss },
      _react2.default.createElement(
        'div',
        { className: 'modal-body' },
        message
      ),
      _react2.default.createElement(
        'div',
        { className: 'modal-footer' },
        showCancel ? _react2.default.createElement(
          _Button2.default,
          { type: 'secondary', onClick: function onClick() {
              return cancel();
            } },
          cancelText
        ) : '',
        _react2.default.createElement(
          _Button2.default,
          { type: 'strong', onClick: function onClick() {
              return proceed();
            } },
          confirmText
        )
      )
    );
  };

  return Dialog;
}(_react2.default.PureComponent);

Dialog.propTypes = {
  show: _react.PropTypes.bool,
  proceed: _react.PropTypes.func,
  cancel: _react.PropTypes.func,
  dismiss: _react.PropTypes.func,
  message: _react.PropTypes.oneOfType([_react.PropTypes.array, _react.PropTypes.element, _react.PropTypes.string]),
  title: _react.PropTypes.string,
  showCancel: _react.PropTypes.bool,
  cancelText: _react.PropTypes.string,
  confirmText: _react.PropTypes.string
};
Dialog.defaultProps = {
  cancelText: 'Cancel',
  confirmText: 'OK'
};
exports.default = Dialog;
module.exports = exports['default'];
