// react-base-ui/src/utils/eventEmitter.js

import clone from 'lodash/clone';
import last from 'lodash/last';

function Events() {
}

function EventListener(fn, context, once) {
  this.fn = fn;
  this.context = context;
  this.once = once || false;
}

class EventEmitter {
  constructor(prefix = '~', topOnly = false) {
    this.prefix = prefix;
    this.topOnly = topOnly;
    this._events = new Events();
    this._eventsCount = 0;
  }

  eventNames() {
    const names = [];
    let events;

    if (this._eventsCount === 0) return names;

    this._events.keys.forEach(name => names.push(this.prefix ? name.slice(1) : name));

    if (Object.getOwnPropertySymbols) {
      return names.concat(Object.getOwnPropertySymbols(events));
    }

    return names;
  }

  listeners(event, exists) {
    const evt = this.prefix ? this.prefix + event : event;
    const available = this._events[evt];

    if (exists) return !!available;
    if (!available) return [];
    if (available.fn) return [available.fn];

    return clone(available);
  }

  emit(event, ...args) {
    const evt = this.prefix ? this.prefix + event : event;

    if (!this._events[evt]) return false;

    let listeners = this._events[evt];
    if (listeners.fn) {
      listeners = [listeners];
    }

    const fire = (listener) => {
      if (listener.once) this._removeListener(event, listener.fn, undefined, true);
      listener.fn.apply(listener.context, args);
    };

    if (this.topOnly) {
      fire(last(listeners));
    } else {
      listeners.forEach(listener => fire(listener));
    }

    return true;
  }

  _addListener(event, fn, context, once = false) {
    const listener = new EventListener(fn, context || this, once);
    const evt = this.prefix ? this.prefix + event : event;

    if (!this._events[evt]) {
      this._events[evt] = listener;
      this._eventsCount = this._eventsCount + 1;
    } else if (!this._events[evt].fn) this._events[evt].push(listener);
    else {
      this._events[evt] = [this._events[evt], listener];
    }

    return this;
  }

  _removeListener(event, fn, context, once) {
    const evt = this.prefix ? this.prefix + event : event;

    if (!this._events[evt]) return this;
    if (!fn) {
      this._eventsCount = this._eventsCount - 1;
      if (this._eventsCount === 0) {
        this._events = new Events();
      } else {
        delete this._events[evt];
      }
      return this;
    }

    let listeners = this._events[evt];
    if (listeners.fn) {
      listeners = [listeners];
    }

    listeners = listeners.filter(listener => (listener.fn !== fn
          || (once && !listener.once)
          || (context && listener.context !== context)));

    if (listeners.length) {
      this._events[evt] = listeners.length === 1 ? listeners[0] : listeners;
    } else {
      this._eventsCount = this._eventsCount - 1;
      if (this._eventsCount === 0) {
        this._events = new Events();
      } else {
        delete this._events[evt];
      }
    }

    return this;
  }

  on(event, fn, context) {
    return this._addListener(event, fn, context);
  }

  once(event, fn, context) {
    return this._addListener(event, fn, context, true);
  }

  off(event, fn, context, once) {
    return this._removeListener(event, fn, context, once);
  }

  isOn(event, fn) {
    const evt = this.prefix ? this.prefix + event : event;

    if (!this._events[evt] || !fn) return false;

    let listeners = this._events[evt];
    if (listeners.fn) {
      listeners = [listeners];
    }

    return listeners.some(listener => (listener.fn === fn));
  }

  removeAllListeners(event) {
    let evt;

    if (event) {
      evt = this.prefix ? this.prefix + event : event;
      if (this._events[evt]) {
        this._eventsCount = this._eventsCount - 1;
        if (this._eventsCount === 0) {
          this._events = new Events();
        } else {
          delete this._events[evt];
        }
      }
    } else {
      this._events = new Events();
      this._eventsCount = 0;
    }

    return this;
  }
}

export default EventEmitter;
