// react-aaui/lib/Carousel.js
'use strict';

exports.__esModule = true;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _taggedTemplateLiteralLoose2 = require('babel-runtime/helpers/taggedTemplateLiteralLoose');

var _taggedTemplateLiteralLoose3 = _interopRequireDefault(_taggedTemplateLiteralLoose2);

var _defineProperties = require('babel-runtime/core-js/object/define-properties');

var _defineProperties2 = _interopRequireDefault(_defineProperties);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _templateObject = (0, _taggedTemplateLiteralLoose3.default)(['aaui-icon-left-arrow\n                              aaui-carousel-left-arrow\n                              ', ''], ['aaui-icon-left-arrow\n                              aaui-carousel-left-arrow\n                              ', '']),
    _templateObject2 = (0, _taggedTemplateLiteralLoose3.default)(['aaui-icon-right-arrow\n                              aaui-carousel-right-arrow\n                              ', ''], ['aaui-icon-right-arrow\n                              aaui-carousel-right-arrow\n                              ', '']);
/* eslint import/extensions: 0, import/no-unresolved: 0 */
/* eslint import/no-extraneous-dependencies: 0, import/no-webpack-loader-syntax: 0 */


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _TweenLite = require('imports?define=>false!gsap/src/uncompressed/TweenLite');

var _TweenLite2 = _interopRequireDefault(_TweenLite);

require('imports?define=>false!gsap/src/uncompressed/plugins/CSSPlugin');

var _utils = require('./shared/utils');

require('./Carousel.less');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var string = _react.PropTypes.string,
    object = _react.PropTypes.object,
    func = _react.PropTypes.func,
    array = _react.PropTypes.array,
    element = _react.PropTypes.element,
    oneOfType = _react.PropTypes.oneOfType,
    number = _react.PropTypes.number,
    any = _react.PropTypes.any;


var carouselPropTypes = {
  className: string,
  style: object,
  children: oneOfType([array, object, func, element]),
  selected: number,
  defaultSelected: number,
  onChange: func
};

var defaultCarouselPropTypes = {
  shown: false
};

var Carousel = function (_PureComponent) {
  (0, _inherits3.default)(Carousel, _PureComponent);

  function Carousel(props) {
    (0, _classCallCheck3.default)(this, Carousel);

    var _this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call(this, props));

    _this.select = function (ind) {
      return function () {
        if (_this.props.selected === undefined) {
          _this.setState({ selected: ind }, function () {
            if (_this.props.onChange) _this.props.onChange(ind);
          });
        } else if (_this.props.onChange) _this.props.onChange(ind);
      };
    };

    _this.state = {
      selected: _this.props.selected != null ? _this.props.selected : _this.props.defaultSelected || 0
    };
    return _this;
  }

  Carousel.prototype.componentDidMount = function componentDidMount() {
    (0, _defineProperties2.default)(this, {
      selected: {
        get: function get() {
          return this.state.selected;
        },
        set: function set(v) {
          if (this.props.selected === undefined) {
            this.setState({ selected: v });
          }
        }
      }
    });

    var slidesCount = _react2.default.Children.count(this.props.children);
    _TweenLite2.default.set(this.slides, { x: '-' + 100 / slidesCount * this.state.selected + '%' });
  };

  /* eslint no-nested-ternary:  */


  Carousel.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    var newState = {
      selected: nextProps.selected !== this.props.selected ? nextProps.selected : nextProps.defaultSelected !== this.props.defaultSelected ? this.state.selected === undefined ? nextProps.defaultSelected : this.state.selected : this.state.selected
    };
    if (newState.selected !== this.state.selected) this.setState(newState);
  };

  Carousel.prototype.componentDidUpdate = function componentDidUpdate(_, _ref) {
    var prevSelected = _ref.selected;

    var selected = this.state.selected;
    if (prevSelected !== selected) {
      var slidesCount = _react2.default.Children.count(this.props.children);
      _TweenLite2.default.to(this.slides, 0.3, { x: '-' + 100 / slidesCount * selected + '%' });
    }
  };

  Carousel.prototype.render = function render() {
    var _this2 = this;

    var _props = this.props,
        className = _props.className,
        style = _props.style,
        children = _props.children;
    var selected = this.state.selected;

    var slidesCount = _react2.default.Children.count(children);
    var slides = _react2.default.Children.map(children, function (slide) {
      return _react2.default.cloneElement(slide, { slidesCount: slidesCount });
    });

    return _react2.default.createElement(
      'div',
      {
        className: 'aaui-carousel ' + (className || ''),
        style: style
      },
      _react2.default.createElement(
        'div',
        {
          ref: function ref(_ref2) {
            _this2.slides = _ref2;
          },
          className: 'aaui-carousel-slides',
          style: { width: 100 * slidesCount + '%' }
        },
        slides
      ),
      _react2.default.createElement(
        'ul',
        { className: 'aaui-carousel-arrows' },
        _react2.default.createElement(
          'li',
          null,
          _react2.default.createElement('span', {
            className: (0, _utils.cls)(_templateObject, selected === 0 ? 'aaui-hidden' : ''),
            onClick: this.select(selected - 1)
          })
        ),
        _react2.default.createElement(
          'li',
          null,
          _react2.default.createElement('span', {
            className: (0, _utils.cls)(_templateObject2, selected === slidesCount - 1 ? 'aaui-hidden' : ''),
            onClick: this.select(selected + 1)
          })
        )
      ),
      _react2.default.createElement(
        'ul',
        { className: 'aaui-carousel-dot' },
        _react2.default.Children.map(children, function (slide, i) {
          return _react2.default.createElement('li', {
            key: i,
            className: i === selected ? 'aaui-carousel-active-dot' : undefined,
            onClick: _this2.select(i)
          });
        })
      )
    );
  };

  return Carousel;
}(_react.PureComponent);

Carousel.displayName = 'AAUICarousel';
Carousel.propTypes = carouselPropTypes;
Carousel.defaultProps = defaultCarouselPropTypes;

var slidePropTypes = {
  className: string,
  style: object,
  children: any,
  slidesCount: number
};

var defaultSlidePropTypes = {
  shown: false
};

/* eslint react/no-multi-comp: 0 */

var Slide = function (_PureComponent2) {
  (0, _inherits3.default)(Slide, _PureComponent2);

  function Slide() {
    (0, _classCallCheck3.default)(this, Slide);
    return (0, _possibleConstructorReturn3.default)(this, _PureComponent2.apply(this, arguments));
  }

  Slide.prototype.render = function render() {
    var _props2 = this.props,
        className = _props2.className,
        style = _props2.style,
        children = _props2.children,
        slidesCount = _props2.slidesCount;


    return _react2.default.createElement(
      'div',
      {
        className: 'aaui-carousel-slide ' + (className || ''),
        style: (0, _assign2.default)({ width: 100 / slidesCount + '%' }, style)
      },
      children
    );
  };

  return Slide;
}(_react.PureComponent);

Slide.displayName = 'AAUICarouselSlide';
Slide.propTypes = slidePropTypes;
Slide.defaultProps = defaultSlidePropTypes;

Carousel.Slide = Slide;

exports.default = Carousel;
module.exports = exports['default'];
