// react-aaui/lib/Checkbox.js
'use strict';

exports.__esModule = true;

var _taggedTemplateLiteralLoose2 = require('babel-runtime/helpers/taggedTemplateLiteralLoose');

var _taggedTemplateLiteralLoose3 = _interopRequireDefault(_taggedTemplateLiteralLoose2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _defineProperties = require('babel-runtime/core-js/object/define-properties');

var _defineProperties2 = _interopRequireDefault(_defineProperties);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _templateObject = (0, _taggedTemplateLiteralLoose3.default)(['checkbox\n          ', '\n          ', '\n          ', ''], ['checkbox\n          ', '\n          ', '\n          ', '']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('active.css/less/components/checkbox.less');

var _utils = require('./shared/utils');

var _keyCodes = require('./shared/keyCodes');

var _keyCodes2 = _interopRequireDefault(_keyCodes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var bool = _react.PropTypes.bool,
    string = _react.PropTypes.string,
    object = _react.PropTypes.object,
    func = _react.PropTypes.func,
    any = _react.PropTypes.any;

var propTypes = {
  size: string,
  children: any,
  className: string,
  style: object,
  disabled: bool,
  checked: bool,
  defaultChecked: bool,
  onChange: func
};
var defaultProps = {};

var Checkbox = function (_PureComponent) {
  (0, _inherits3.default)(Checkbox, _PureComponent);

  function Checkbox(props) {
    (0, _classCallCheck3.default)(this, Checkbox);

    var _this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call(this, props));

    _this.setWrappedComponentInstance = function (input) {
      _this.input = input;
    };

    _this.handleChange = function (e) {
      e.persist();

      _this.setState({
        checked: e.target.checked
      }, function () {
        if (_this.props.onChange) {
          _this.props.onChange(e);
        }
      });
    };

    _this.handleKeyDown = function (e) {
      e.persist();

      // Prevent scrolling when pressing any scroll key (e.g. down arrow, spacebar) when on a widget
      switch (e.keyCode) {
        case _keyCodes2.default.SPACE:
          // SPACEBAR
          e.preventDefault();

          _this.setState({
            checked: !_this.state.checked
          }, function () {
            if (_this.props.onChange) {
              // We must change the `target` property
              // in order to feed the `onChange` callback when triggering on the container
              _this.props.onChange((0, _extends3.default)({}, e, {
                target: _this.input
              }));
            }
          });
          break;
        case _keyCodes2.default.UPARROW: // UPARROW
        case _keyCodes2.default.DOWNARROW:
          // DOWNARROW
          e.preventDefault();
          break;
        default:
          break;
      }
    };

    var _props$defaultChecked = props.defaultChecked,
        defaultChecked = _props$defaultChecked === undefined ? false : _props$defaultChecked,
        _props$checked = props.checked,
        checked = _props$checked === undefined ? false : _props$checked;


    _this.state = {
      checked: defaultChecked || checked
    };
    return _this;
  }

  Checkbox.prototype.componentDidMount = function componentDidMount() {
    (0, _defineProperties2.default)(this, {
      checked: {
        get: function get() {
          return this.state.checked;
        },
        set: function set(v) {
          var _this2 = this;

          this.setState({
            checked: !!v
          }, function () {
            if (_this2.props.onChange) {
              _this2.props.onChange({
                target: _this2.input
              });
            }
          });
        }
      }
    });
  };

  Checkbox.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    if (nextProps.checked !== this.props.checked) {
      this.setState({
        checked: nextProps.checked
      });
    }
  };

  Checkbox.prototype.render = function render() {
    /* eslint-disable no-unused-vars */
    var _props = this.props,
        size = _props.size,
        children = _props.children,
        className = _props.className,
        style = _props.style,
        disabled = _props.disabled,
        defaultChecked = _props.defaultChecked,
        rest = (0, _objectWithoutProperties3.default)(_props, ['size', 'children', 'className', 'style', 'disabled', 'defaultChecked']);

    /* eslint-disable jsx-a11y/label-has-for */

    return _react2.default.createElement(
      'label',
      {
        className: (0, _utils.cls)(_templateObject, !children ? 'checkbox--empty' : '', size ? 'checkbox--' + size : '', className || ''),
        style: style,
        onKeyDown: this.handleKeyDown
      },
      _react2.default.createElement('input', (0, _extends3.default)({
        ref: this.setWrappedComponentInstance,
        disabled: disabled,
        type: 'checkbox',
        'aria-hidden': true
      }, rest, {
        checked: this.state.checked,
        onChange: this.handleChange
      })),
      _react2.default.createElement(
        'span',
        {
          role: 'checkbox',
          className: 'checkbox__text',
          tabIndex: disabled ? -1 : 0,
          'aria-disabled': disabled,
          'aria-checked': this.state.checked
        },
        children
      )
    );
  };

  return Checkbox;
}(_react.PureComponent);

exports.default = Checkbox;


Checkbox.displayName = 'AAUICheckbox';
Checkbox.propTypes = propTypes;
Checkbox.defaultProps = defaultProps;
module.exports = exports['default'];
