// react-aaui/lib/MoreButton.js
'use strict';

exports.__esModule = true;

var _taggedTemplateLiteralLoose2 = require('babel-runtime/helpers/taggedTemplateLiteralLoose');

var _taggedTemplateLiteralLoose3 = _interopRequireDefault(_taggedTemplateLiteralLoose2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _templateObject = (0, _taggedTemplateLiteralLoose3.default)(['more-button ', ''], ['more-button ', '']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('active.css/less/components/more-button.less');

var _utils = require('./shared/utils');

var _dataAccess = require('./shared/data-access');

var da = _interopRequireWildcard(_dataAccess);

var _Dropdown = require('./Dropdown');

var _Dropdown2 = _interopRequireDefault(_Dropdown);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var string = _react.PropTypes.string,
    object = _react.PropTypes.object,
    bool = _react.PropTypes.bool;


var propTypes = {
  className: string,
  title: string,
  data: object,
  style: object,
  filter: bool,
  filterPlaceholder: string,
  disabled: bool
};

var defaultProps = {
  title: 'More'
};

var MoreButton = function (_PureComponent) {
  (0, _inherits3.default)(MoreButton, _PureComponent);

  function MoreButton() {
    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, MoreButton);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call.apply(_PureComponent, [this].concat(args))), _this), _this.invokeAction = function (_ref) {
      var k = _ref.value;

      da.get(_this.props.data, k)();
    }, _this.render = function () {
      var _this$props = _this.props,
          className = _this$props.className,
          style = _this$props.style,
          title = _this$props.title,
          filter = _this$props.filter,
          filterPlaceholder = _this$props.filterPlaceholder,
          disabled = _this$props.disabled;

      var data = da.keys(_this.props.data).map(function (k) {
        return { text: k, value: k };
      });

      return _react2.default.createElement(_Dropdown2.default, {
        className: (0, _utils.cls)(_templateObject, className || ''),
        style: style,
        placeholder: title,
        filter: filter,
        filterPlaceholder: filterPlaceholder,
        data: data,
        value: '',
        disabled: disabled,
        isMoreButton: true,
        onChange: _this.invokeAction
      });
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  return MoreButton;
}(_react.PureComponent);

exports.default = MoreButton;


MoreButton.displayName = 'AAUIMoreButton';
MoreButton.propTypes = propTypes;
MoreButton.defaultProps = defaultProps;
module.exports = exports['default'];
