// react-aaui/lib/RadioGroup.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('active.css/less/components/radio-group.less');

var _Radio = require('./Radio');

var _Radio2 = _interopRequireDefault(_Radio);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var string = _react.PropTypes.string,
    func = _react.PropTypes.func,
    node = _react.PropTypes.node;

var propTypes = {
  children: node.isRequired,
  className: string,
  name: string,
  onChange: func,
  selected: string,
  size: string
};
var defaultProps = {
  className: 'radio-group'
};

var RadioGroup = function (_PureComponent) {
  (0, _inherits3.default)(RadioGroup, _PureComponent);

  function RadioGroup() {
    (0, _classCallCheck3.default)(this, RadioGroup);
    return (0, _possibleConstructorReturn3.default)(this, _PureComponent.apply(this, arguments));
  }

  RadioGroup.prototype.render = function render() {
    var _props = this.props,
        name = _props.name,
        children = _props.children,
        className = _props.className,
        onChange = _props.onChange,
        selected = _props.selected,
        size = _props.size,
        rest = (0, _objectWithoutProperties3.default)(_props, ['name', 'children', 'className', 'onChange', 'selected', 'size']);


    return _react2.default.createElement(
      'div',
      (0, _extends3.default)({
        className: className
      }, rest),
      _react2.default.Children.map(children, function (child) {
        if (child.type === _Radio2.default) {
          return _react2.default.cloneElement(child, {
            name: name,
            onChange: onChange,
            checked: child.props.value === selected,
            size: size
          });
        }

        return child;
      })
    );
  };

  return RadioGroup;
}(_react.PureComponent);

exports.default = RadioGroup;


RadioGroup.displayName = 'RadioGroup';
RadioGroup.propTypes = propTypes;
RadioGroup.defaultProps = defaultProps;
module.exports = exports['default'];
