// react-aaui/lib/Sidebar.js
'use strict';

exports.__esModule = true;

var _taggedTemplateLiteralLoose2 = require('babel-runtime/helpers/taggedTemplateLiteralLoose');

var _taggedTemplateLiteralLoose3 = _interopRequireDefault(_taggedTemplateLiteralLoose2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _templateObject = (0, _taggedTemplateLiteralLoose3.default)(['sidebar\n          ', '\n          ', ''], ['sidebar\n          ', '\n          ', '']),
    _templateObject2 = (0, _taggedTemplateLiteralLoose3.default)(['', '\n                      ', ''], ['', '\n                      ', '']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('active.css/less/components/sidebar.less');

var _utils = require('./shared/utils');

var _dataAccess = require('./shared/data-access');

var da = _interopRequireWildcard(_dataAccess);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var bool = _react.PropTypes.bool,
    string = _react.PropTypes.string,
    number = _react.PropTypes.number,
    object = _react.PropTypes.object,
    array = _react.PropTypes.array,
    element = _react.PropTypes.element,
    oneOfType = _react.PropTypes.oneOfType;


var propTypes = {
  className: string,
  style: object,
  data: array,
  links: bool,
  steps: bool,
  active: number,
  past: number,
  priority: string,
  title: string,
  linkElement: oneOfType([string, element])
};

var defaultProps = {
  priority: '5',
  steps: false,
  linkElement: 'a'
};

var Sidebar = function (_PureComponent) {
  (0, _inherits3.default)(Sidebar, _PureComponent);

  function Sidebar() {
    (0, _classCallCheck3.default)(this, Sidebar);
    return (0, _possibleConstructorReturn3.default)(this, _PureComponent.apply(this, arguments));
  }

  Sidebar.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        style = _props.style,
        _props$data = _props.data,
        data = _props$data === undefined ? [] : _props$data,
        steps = _props.steps,
        active = _props.active,
        past = _props.past,
        priority = _props.priority,
        title = _props.title,
        links = _props.links,
        A = _props.linkElement;


    var HeaderElement = 'h' + priority;

    return _react2.default.createElement(
      'aside',
      {
        className: (0, _utils.cls)(_templateObject, links ? 'sidebar--links' : '', className || ''),
        style: style
      },
      _react2.default.createElement(
        HeaderElement,
        null,
        title
      ),
      _react2.default.createElement(
        'ul',
        null,
        data.map(function (item, i) {
          var stepNum = steps ? i + 1 + '.' : undefined;

          return _react2.default.createElement(
            'li',
            {
              key: i
            },
            _react2.default.createElement(
              A,
              {
                className: (0, _utils.cls)(_templateObject2, i === active ? 'active' : '', i === past ? 'past' : ''),
                href: da.get(item, 'href')
              },
              stepNum,
              ' ',
              da.get(item, 'text')
            )
          );
        })
      )
    );
  };

  return Sidebar;
}(_react.PureComponent);

exports.default = Sidebar;


Sidebar.displayName = 'AAUISidebar';
Sidebar.propTypes = propTypes;
Sidebar.defaultProps = defaultProps;
module.exports = exports['default'];
