// react-aaui/lib/Tabs.js
'use strict';

exports.__esModule = true;

var _defineProperties = require('babel-runtime/core-js/object/define-properties');

var _defineProperties2 = _interopRequireDefault(_defineProperties);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('active.css/less/components/tab.less');

var _getContext = require('./shared/getContext');

var _getContext2 = _interopRequireDefault(_getContext);

var _TabsHeader = require('./TabsHeader');

var _TabsHeader2 = _interopRequireDefault(_TabsHeader);

var _TabsTitle = require('./TabsTitle');

var _TabsTitle2 = _interopRequireDefault(_TabsTitle);

var _TabsContainer = require('./TabsContainer');

var _TabsContainer2 = _interopRequireDefault(_TabsContainer);

var _types = require('./shared/types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var string = _react.PropTypes.string,
    object = _react.PropTypes.object,
    func = _react.PropTypes.func,
    any = _react.PropTypes.any;


var propTypes = {
  className: string,
  style: object,
  selected: string,
  defaultSelected: string,
  onChange: func,
  children: any
};

var childContextTypes = {
  aauiTabsAPI: _types.tabsAPIShape
};

var Tabs = function (_PureComponent) {
  (0, _inherits3.default)(Tabs, _PureComponent);

  function Tabs(props) {
    (0, _classCallCheck3.default)(this, Tabs);

    var _this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call(this, props));

    _this.getSelected = function () {
      return _this.state.selected;
    };

    _this.select = function (name) {
      if (_this.props.selected === undefined) {
        _this.setState({
          selected: name
        }, function () {
          if (_this.props.onChange) _this.props.onChange(name);
        });
      } else if (_this.props.onChange) _this.props.onChange(name);
    };

    _this.state = {
      selected: props.selected || props.defaultSelected
    };
    return _this;
  }

  Tabs.prototype.getChildContext = function getChildContext() {
    return {
      aauiTabsAPI: {
        select: this.select,
        getSelected: this.getSelected
      }
    };
  };

  Tabs.prototype.componentDidMount = function componentDidMount() {
    var _this2 = this;

    (0, _defineProperties2.default)(this, {
      selected: {
        get: function get() {
          return this.state.selected;
        },
        set: function set(v) {
          if (this.props.selected === undefined) {
            this.setState({
              selected: v
            });
          }
        }
      }
    });

    if (this.state.selected == null) {
      _react2.default.Children.forEach(this.props.children, function (child) {
        if (child.type === Tabs.Container) {
          _this2.setState({
            selected: child.props.name
          });
        }
      });
    }
  };

  Tabs.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    var defaultSelected = nextProps.defaultSelected !== this.props.defaultSelected ? nextProps.defaultSelected : this.state.selected;
    var newState = {
      selected: nextProps.selected !== this.props.selected ? nextProps.selected : defaultSelected
    };

    if (newState.selected !== this.state.selected) this.setState(newState);
  };

  Tabs.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        style = _props.style,
        children = _props.children;


    return _react2.default.createElement(
      'div',
      {
        className: className || '',
        style: style
      },
      children
    );
  };

  return Tabs;
}(_react.PureComponent);

exports.default = Tabs;


Tabs.displayName = 'AAUITabs';
Tabs.propTypes = propTypes;
Tabs.childContextTypes = childContextTypes;

Tabs.Header = _TabsHeader2.default;
Tabs.Title = (0, _getContext2.default)(childContextTypes)(_TabsTitle2.default);
Tabs.Container = (0, _getContext2.default)(childContextTypes)(_TabsContainer2.default);
module.exports = exports['default'];
