// react-aaui/lib/Tag.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _taggedTemplateLiteralLoose2 = require('babel-runtime/helpers/taggedTemplateLiteralLoose');

var _taggedTemplateLiteralLoose3 = _interopRequireDefault(_taggedTemplateLiteralLoose2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _templateObject = (0, _taggedTemplateLiteralLoose3.default)(['tag-input\n        ', ''], ['tag-input\n        ', '']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('active.css/less/components/tag.less');

var _utils = require('./shared/utils');

var _keyCodes = require('./shared/keyCodes');

var _keyCodes2 = _interopRequireDefault(_keyCodes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var string = _react.PropTypes.string,
    func = _react.PropTypes.func,
    bool = _react.PropTypes.bool;

var propTypes = {
  className: string,
  value: string,
  onChange: func,
  onCancel: func,
  onClose: func,
  editMode: bool,
  isNew: bool
};
var defaultProps = {
  editMode: false
};

var Tag = function (_PureComponent) {
  (0, _inherits3.default)(Tag, _PureComponent);

  function Tag(props) {
    (0, _classCallCheck3.default)(this, Tag);

    var _this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call(this, props));

    _this.handleKeyDown = function (e) {
      var target = e.target;

      switch (e.keyCode) {
        case _keyCodes2.default.ESCAPE:
          e.preventDefault();
          _this.setState({
            isInputCancel: true,
            isTabEnter: false
          }, function () {
            target.blur();
          });
          break;
        case _keyCodes2.default.TAB:
        case _keyCodes2.default.ENTER:
          e.preventDefault();

          _this.setState({
            isTabEnter: true
          }, function () {
            if (target) {
              target.blur();
            }
          });
          break;
        default:
          break;
      }
    };

    _this.handleBlur = function (e) {
      var value = e.target.value;
      _this.setState({
        isInput: false
      }, function () {
        if (!_this.state.isInputCancel) {
          var onChange = _this.props.onChange;
          if (onChange) onChange(value, _this.state.isTabEnter);
        } else {
          var onCancel = _this.props.onCancel;
          if (onCancel) onCancel();
        }
      });
    };

    _this.toInput = function (e) {
      e && e.stopPropagation && e.stopPropagation();

      _this.setState({
        isInput: true,
        isInputCancel: false
      }, function () {
        var input = _this.input;

        if (input) {
          input.focus();
          input.select();
        }
      });
    };

    _this.state = {
      isInput: false,
      isTabEnter: false
    };
    return _this;
  }

  Tag.prototype.render = function render() {
    var _this2 = this;

    /* eslint-disable no-unused-vars */
    var _props = this.props,
        onCancel = _props.onCancel,
        className = _props.className,
        editMode = _props.editMode,
        onClose = _props.onClose,
        onChange = _props.onChange,
        value = _props.value,
        isNew = _props.isNew,
        rest = (0, _objectWithoutProperties3.default)(_props, ['onCancel', 'className', 'editMode', 'onClose', 'onChange', 'value', 'isNew']);
    var isInput = this.state.isInput;


    if (!editMode) {
      return _react2.default.createElement(
        'span',
        { className: 'badge' },
        value
      );
    }

    return _react2.default.createElement(
      'span',
      (0, _extends3.default)({
        className: (0, _utils.cls)(_templateObject, className || '')
      }, rest),
      isInput ? _react2.default.createElement('input', {
        className: 'input',
        ref: function ref(input) {
          _this2.input = input;
        },
        placeholder: isNew ? value : undefined,
        defaultValue: isNew ? undefined : value,
        onKeyDown: this.handleKeyDown,
        onBlur: this.handleBlur
      }) : _react2.default.createElement(
        'span',
        { className: 'badge', onClick: onClose },
        _react2.default.createElement(
          'span',
          { onClick: this.toInput },
          value
        )
      )
    );
  };

  return Tag;
}(_react.PureComponent);

exports.default = Tag;


Tag.displayName = 'AAUITag';
Tag.propTypes = propTypes;
Tag.defaultProps = defaultProps;
module.exports = exports['default'];
