// react-aaui/lib/Alert.js
'use strict';

exports.__esModule = true;

var _taggedTemplateLiteralLoose2 = require('babel-runtime/helpers/taggedTemplateLiteralLoose');

var _taggedTemplateLiteralLoose3 = _interopRequireDefault(_taggedTemplateLiteralLoose2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _templateObject = (0, _taggedTemplateLiteralLoose3.default)(['alert\n        alert-', '\n        ', '\n        ', ''], ['alert\n        alert-', '\n        ', '\n        ', '']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('active.css/less/components/alert.less');

var _utils = require('./shared/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var bool = _react.PropTypes.bool,
    string = _react.PropTypes.string,
    func = _react.PropTypes.func,
    object = _react.PropTypes.object,
    any = _react.PropTypes.any,
    oneOf = _react.PropTypes.oneOf;


var icons = {
  success: 'success',
  warning: 'warn',
  danger: 'close-circle',
  info: 'info'
};

var propTypes = {
  type: oneOf(['success', 'warning', 'danger', 'info']).isRequired,
  className: string,
  style: object,
  noClose: bool,
  onClose: func,
  children: any
};

var defaultProps = {
  type: 'info',
  noClose: false
};

var Alert = function (_PureComponent) {
  (0, _inherits3.default)(Alert, _PureComponent);

  function Alert() {
    (0, _classCallCheck3.default)(this, Alert);
    return (0, _possibleConstructorReturn3.default)(this, _PureComponent.apply(this, arguments));
  }

  Alert.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        style = _props.style,
        type = _props.type,
        noClose = _props.noClose,
        onClose = _props.onClose,
        children = _props.children;


    return _react2.default.createElement(
      'div',
      {
        className: (0, _utils.cls)(_templateObject, type, className || '', noClose ? undefined : 'alert-dismissable'),
        style: style
      },
      _react2.default.createElement('span', { className: 'icon-' + icons[type] }),
      children,
      noClose ? undefined : _react2.default.createElement(
        'button',
        { type: 'button', className: 'close', onClick: onClose },
        '\xD7'
      )
    );
  };

  return Alert;
}(_react.PureComponent);

exports.default = Alert;


Alert.displayName = 'AAUIAlert';
Alert.propTypes = propTypes;
Alert.defaultProps = defaultProps;
module.exports = exports['default'];
