// react-aaui/lib/Button.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _taggedTemplateLiteralLoose2 = require('babel-runtime/helpers/taggedTemplateLiteralLoose');

var _taggedTemplateLiteralLoose3 = _interopRequireDefault(_taggedTemplateLiteralLoose2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _templateObject = (0, _taggedTemplateLiteralLoose3.default)(['btn\n          btn-', '\n          ', '\n          ', ''], ['btn\n          btn-', '\n          ', '\n          ', '']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('active.css/less/components/button.less');

var _utils = require('./shared/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var bool = _react.PropTypes.bool,
    string = _react.PropTypes.string,
    any = _react.PropTypes.any;


var propTypes = {
  noSubmit: bool,
  type: string.isRequired,
  size: string,
  className: string,
  children: any
};

var defaultProps = {
  noSubmit: false,
  type: 'secondary'
};

var Button = function (_PureComponent) {
  (0, _inherits3.default)(Button, _PureComponent);

  function Button() {
    (0, _classCallCheck3.default)(this, Button);
    return (0, _possibleConstructorReturn3.default)(this, _PureComponent.apply(this, arguments));
  }

  Button.prototype.render = function render() {
    var _props = this.props,
        noSubmit = _props.noSubmit,
        type = _props.type,
        size = _props.size,
        children = _props.children,
        className = _props.className,
        rest = (0, _objectWithoutProperties3.default)(_props, ['noSubmit', 'type', 'size', 'children', 'className']);


    return _react2.default.createElement(
      'button',
      (0, _extends3.default)({}, rest, {
        type: noSubmit ? 'button' : 'submit',
        className: (0, _utils.cls)(_templateObject, type, size ? 'btn--' + size : '', className || '')
      }),
      children
    );
  };

  return Button;
}(_react.PureComponent);

exports.default = Button;


Button.displayName = 'AAUIButton';
Button.propTypes = propTypes;
Button.defaultProps = defaultProps;
module.exports = exports['default'];
