// react-aaui/lib/Infobar.js
'use strict';

exports.__esModule = true;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('active.css/less/components/alert.less');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var bool = _react.PropTypes.bool,
    string = _react.PropTypes.string,
    object = _react.PropTypes.object,
    func = _react.PropTypes.func,
    any = _react.PropTypes.any;


var propTypes = {
  className: string,
  style: object,
  children: any,
  noClose: bool,
  onClose: func
};

var defaultProps = {
  noClose: false
};

var Infobar = function (_PureComponent) {
  (0, _inherits3.default)(Infobar, _PureComponent);

  function Infobar() {
    (0, _classCallCheck3.default)(this, Infobar);
    return (0, _possibleConstructorReturn3.default)(this, _PureComponent.apply(this, arguments));
  }

  Infobar.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        style = _props.style,
        children = _props.children,
        noClose = _props.noClose,
        onClose = _props.onClose;


    return _react2.default.createElement(
      'div',
      {
        className: 'infobar ' + (className || ''),
        style: style
      },
      children,
      noClose || _react2.default.createElement(
        'button',
        { type: 'button', className: 'close', onClick: onClose },
        '\xD7'
      )
    );
  };

  return Infobar;
}(_react.PureComponent);

exports.default = Infobar;


Infobar.propTypes = propTypes;
Infobar.defaultProps = defaultProps;
Infobar.displayName = 'AAUIInfobar';
module.exports = exports['default'];
