// react-aaui/lib/Input.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _taggedTemplateLiteralLoose2 = require('babel-runtime/helpers/taggedTemplateLiteralLoose');

var _taggedTemplateLiteralLoose3 = _interopRequireDefault(_taggedTemplateLiteralLoose2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _defineProperties = require('babel-runtime/core-js/object/define-properties');

var _defineProperties2 = _interopRequireDefault(_defineProperties);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _templateObject = (0, _taggedTemplateLiteralLoose3.default)(['input-group\n        ', '\n        ', '\n        ', ''], ['input-group\n        ', '\n        ', '\n        ', '']),
    _templateObject2 = (0, _taggedTemplateLiteralLoose3.default)(['icon ', ''], ['icon ', '']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('active.css/less/components/input.less');

require('active.css/less/components/input-group.less');

var _utils = require('./shared/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var bool = _react.PropTypes.bool,
    string = _react.PropTypes.string,
    func = _react.PropTypes.func,
    object = _react.PropTypes.object,
    any = _react.PropTypes.any;


var propTypes = {
  value: string,
  defaultValue: string,
  type: string.isRequired,
  size: string,
  preIcon: string,
  preText: string,
  postText: string,
  errored: bool,
  disabled: bool,
  className: string,
  style: object,
  onChange: func,
  children: any
};

var defaultProps = {
  type: 'text'
};

var Input = function (_PureComponent) {
  (0, _inherits3.default)(Input, _PureComponent);

  function Input(props) {
    (0, _classCallCheck3.default)(this, Input);

    var _this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call(this, props));

    _this.setWrappedComponentInstance = function (input) {
      _this.input = input;
    };

    _this.handleChange = function (e) {
      // If you want to access the event properties in an asynchronous way,
      // you should call `event.persist()`` on the event,
      // which will remove the synthetic event from the pool
      // and allow references to the event to be retained by user code.
      e.persist();

      _this.setState({
        value: e.target.value
      }, function () {
        if (_this.props.onChange) {
          _this.props.onChange(e);
        }
      });
    };

    var _props$value = props.value,
        value = _props$value === undefined ? '' : _props$value,
        _props$defaultValue = props.defaultValue,
        defaultValue = _props$defaultValue === undefined ? '' : _props$defaultValue;


    _this.state = {
      value: value || defaultValue
    };
    return _this;
  }

  Input.prototype.componentDidMount = function componentDidMount() {
    (0, _defineProperties2.default)(this, {
      value: {
        get: function get() {
          return this.input.value;
        },
        set: function set(v) {
          if (this.props.value === undefined) {
            this.input.value = v;
          }
        }
      }
    });
  };

  // Data displayed by a controlled component is received through props passed down
  // from it's parent/container component.
  // Sync props -> state


  Input.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    if (nextProps.value !== this.props.value) {
      this.setState({
        value: nextProps.value
      });
    }
  };

  // Controlled components have functions to govern the data going into them
  // on every onChange event, rather than grabbing the data only once,
  // e.g. when a user clicks a submit button.
  // This 'governed' data is then saved to state


  Input.prototype.render = function render() {
    var _props = this.props,
        type = _props.type,
        size = _props.size,
        preIcon = _props.preIcon,
        preText = _props.preText,
        postText = _props.postText,
        errored = _props.errored,
        disabled = _props.disabled,
        className = _props.className,
        style = _props.style,
        rest = (0, _objectWithoutProperties3.default)(_props, ['type', 'size', 'preIcon', 'preText', 'postText', 'errored', 'disabled', 'className', 'style']);


    return _react2.default.createElement(
      'div',
      {
        className: (0, _utils.cls)(_templateObject, size ? 'input-group--' + size : '', errored ? 'input-group--error' : '', className || ''),
        style: style
      },
      preIcon || preText ? _react2.default.createElement(
        'span',
        {
          className: 'input-group__item'
        },
        preIcon && _react2.default.createElement('i', { className: (0, _utils.cls)(_templateObject2, preIcon) }),
        preText
      ) : undefined,
      _react2.default.createElement('input', (0, _extends3.default)({}, rest, {
        type: type,
        disabled: disabled,
        className: 'input input-group__field',
        ref: this.setWrappedComponentInstance,
        value: this.state.value,
        onChange: this.handleChange
      })),
      postText ? _react2.default.createElement(
        'span',
        { className: 'input-group__item' },
        postText
      ) : undefined
    );
  };

  return Input;
}(_react.PureComponent);

exports.default = Input;


Input.displayName = 'AAUIInput';
Input.propTypes = propTypes;
Input.defaultProps = defaultProps;
module.exports = exports['default'];
