// react-aaui/lib/Modal.js
'use strict';

exports.__esModule = true;

var _taggedTemplateLiteralLoose2 = require('babel-runtime/helpers/taggedTemplateLiteralLoose');

var _taggedTemplateLiteralLoose3 = _interopRequireDefault(_taggedTemplateLiteralLoose2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _templateObject = (0, _taggedTemplateLiteralLoose3.default)(['modal ', ' ', ''], ['modal ', ' ', '']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('active.css/less/components/modal.less');

var _utils = require('./shared/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var string = _react.PropTypes.string,
    bool = _react.PropTypes.bool,
    object = _react.PropTypes.object,
    func = _react.PropTypes.func,
    array = _react.PropTypes.array,
    element = _react.PropTypes.element,
    oneOfType = _react.PropTypes.oneOfType;


var propTypes = {
  className: string,
  onClose: func,
  title: string,
  style: object,
  shown: bool,
  children: oneOfType([array, object, func, element])
};

var defaultProps = {
  shown: false
};

var Modal = function (_PureComponent) {
  (0, _inherits3.default)(Modal, _PureComponent);

  function Modal() {
    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, Modal);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call.apply(_PureComponent, [this].concat(args))), _this), _this.componentDidMount = function () {
      if (_this.props.shown) {
        _this.fixBody();
      }
    }, _this.componentDidUpdate = function (_ref) {
      var shown = _ref.shown;

      if (shown !== _this.props.shown) {
        if (_this.props.shown) {
          _this.fixBody();
        } else {
          _this.unfixBody();
        }
      }
    }, _this.fixBody = function () {
      var bodyStyles = document.body.style;
      var scrollbarWidth = window.innerWidth - document.documentElement.clientWidth;

      _this.bodyOverflowX = bodyStyles.overflowX;
      _this.bodyOverflowY = bodyStyles.overflowY;
      _this.bodyPaddingRight = bodyStyles.paddingRight;

      bodyStyles.overflowX = bodyStyles.overflowY = 'hidden';
      if (scrollbarWidth > 0) {
        var paddingRight = getComputedStyle(document.body).paddingRight;
        _this.root.style.paddingRight = bodyStyles.paddingRight = Number(paddingRight) + Number(scrollbarWidth) + 'px';
      }
    }, _this.unfixBody = function () {
      var bodyStyles = document.body.style;
      bodyStyles.overflowX = _this.bodyOverflowX;
      bodyStyles.overflowY = _this.bodyOverflowY;
      bodyStyles.paddingRight = _this.bodyPaddingRight;
      _this.root.style.paddingRight = '';
    }, _this.render = function () {
      var _this$props = _this.props,
          className = _this$props.className,
          style = _this$props.style,
          title = _this$props.title,
          onClose = _this$props.onClose,
          children = _this$props.children;


      return _react2.default.createElement(
        'div',
        { className: 'modal-wrap' },
        _react2.default.createElement(
          'div',
          { ref: function ref(_ref3) {
              _this.root = _ref3;
            }, className: (0, _utils.cls)(_templateObject, _this.props.shown ? 'is-open' : '', className || ''), style: style },
          _react2.default.createElement(
            'section',
            { ref: function ref(_ref2) {
                _this.box = _ref2;
              }, className: 'modal-box' },
            _react2.default.createElement(
              'header',
              { className: 'modal-header' },
              _react2.default.createElement(
                'h3',
                { className: 'modal-title' },
                title
              ),
              _react2.default.createElement('span', { onClick: onClose, className: 'icon-close modal-close' })
            ),
            children
          )
        ),
        _react2.default.createElement('div', { className: 'modal-mask' })
      );
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  return Modal;
}(_react.PureComponent);

exports.default = Modal;


Modal.displayName = 'AAUIModal';
Modal.propTypes = propTypes;
Modal.defaultProps = defaultProps;
module.exports = exports['default'];
