// react-aaui/lib/Pagination.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _taggedTemplateLiteralLoose2 = require('babel-runtime/helpers/taggedTemplateLiteralLoose');

var _taggedTemplateLiteralLoose3 = _interopRequireDefault(_taggedTemplateLiteralLoose2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _templateObject = (0, _taggedTemplateLiteralLoose3.default)(['', '\n                       ', '\n                       ', ''], ['', '\n                       ', '\n                       ', '']),
    _templateObject2 = (0, _taggedTemplateLiteralLoose3.default)(['', ''], ['', '']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('active.css/less/components/pagination.less');

var _utils = require('./shared/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function pageData(type, theme, total, current, around, url, roundText, startText, endText) {
  function pageLink(page) {
    if (typeof url === 'function') {
      return url(page);
    }

    var pageLinkVar = typeof url === 'string' ? url : '';
    if (pageLinkVar.indexOf('?') > -1) {
      pageLinkVar += '&page';
    } else {
      pageLinkVar += '?page';
    }
    return pageLinkVar + '=' + page;
  }

  var start = 1;
  var end = total;
  var aroundStart = Math.max(current - around, start + 1);
  var aroundEnd = Math.min(current + around, end - 1);
  var data = [];

  switch (type) {
    case 'pager':
      if (current > start) {
        data.push({ text: startText, href: pageLink(current - 1) });
      }
      if (current < end) {
        data.push({ text: endText, href: pageLink(current + 1) });
      }
      break;
    case 'pagination':
      if (current > start) {
        data.push({ text: startText, href: pageLink(current - 1) });
      }
      /* eslint prefer-template: 0 */
      data.push({ text: '' + start, href: pageLink(start) });
      if (aroundStart - start > 1) {
        data.push({ text: roundText });
      }
      for (var i = aroundStart; i < current; i += 1) {
        data.push({ text: '' + i, href: pageLink(i) });
      }
      for (var _i = current !== start ? current : current + 1; _i <= aroundEnd; _i += 1) {
        data.push({ text: '' + _i, href: pageLink(_i) });
      }
      if (end - aroundEnd > 1) {
        data.push({ text: roundText });
      }
      data.push({ text: '' + end, href: pageLink(end) });
      if (current < end) {
        data.push({ text: endText, href: pageLink(current + 1) });
      }
      break;
    default:
      break;
  }

  return data;
}

var number = _react.PropTypes.number,
    string = _react.PropTypes.string,
    object = _react.PropTypes.object,
    oneOfType = _react.PropTypes.oneOfType;


var propTypes = {
  className: string,
  type: string,
  theme: string,
  total: number,
  current: number,
  around: number,
  endText: oneOfType([object, string]),
  linkElement: string,
  startText: oneOfType([object, string]),
  roundText: oneOfType([object, string]),
  url: string
};

var defaultProps = {
  type: 'pagination',
  url: '',
  around: 3,
  linkElement: 'a',
  roundText: '...',
  startText: _react2.default.createElement('span', { className: 'icon-uniF053' }),
  endText: _react2.default.createElement('span', { className: 'icon-uniF054' })
};

var Pagination = function (_PureComponent) {
  (0, _inherits3.default)(Pagination, _PureComponent);

  function Pagination() {
    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, Pagination);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call.apply(_PureComponent, [this].concat(args))), _this), _this.render = function () {
      var _this$props = _this.props,
          type = _this$props.type,
          theme = _this$props.theme,
          total = _this$props.total,
          current = _this$props.current,
          around = _this$props.around,
          url = _this$props.url,
          roundText = _this$props.roundText,
          startText = _this$props.startText,
          endText = _this$props.endText,
          A = _this$props.linkElement,
          className = _this$props.className,
          attrs = (0, _objectWithoutProperties3.default)(_this$props, ['type', 'theme', 'total', 'current', 'around', 'url', 'roundText', 'startText', 'endText', 'linkElement', 'className']);


      return _react2.default.createElement(
        'ul',
        (0, _extends3.default)({
          className: (0, _utils.cls)(_templateObject, type, type === 'pagination' && theme ? 'pagination--' + theme : '', className || '')
        }, attrs),
        pageData(type, theme, total, current, around, url, roundText, startText, endText).map(function (_ref, i) {
          var text = _ref.text,
              href = _ref.href;

          if (href && +text !== current) {
            return _react2.default.createElement(
              'li',
              { key: i, className: (0, _utils.cls)(_templateObject2, +text === current ? 'active' : '') },
              _react2.default.createElement(
                A,
                { key: i, href: href },
                text
              )
            );
          } else {
            /* eslint no-else-return: 0 */
            return _react2.default.createElement(
              'li',
              {
                key: i,
                className: (0, _utils.cls)(_templateObject2, +text === current ? 'active' : '')
              },
              _react2.default.createElement(
                'span',
                { key: i },
                text
              )
            );
          }
        })
      );
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  return Pagination;
}(_react.PureComponent);

exports.default = Pagination;


Pagination.displayName = 'AAUIPagination';
Pagination.propTypes = propTypes;
Pagination.defaultProps = defaultProps;
module.exports = exports['default'];
