// react-aaui/lib/SearchInput.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _taggedTemplateLiteralLoose2 = require('babel-runtime/helpers/taggedTemplateLiteralLoose');

var _taggedTemplateLiteralLoose3 = _interopRequireDefault(_taggedTemplateLiteralLoose2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _templateObject = (0, _taggedTemplateLiteralLoose3.default)(['\n          search-box\n          ', '\n        '], ['\n          search-box\n          ', '\n        ']),
    _templateObject2 = (0, _taggedTemplateLiteralLoose3.default)(['\n            search-box__input-addon-group\n            ', '\n          '], ['\n            search-box__input-addon-group\n            ', '\n          ']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('active.css/less/icon.less');

require('active.css/less/components/input-group.less');

var _utils = require('./shared/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var string = _react.PropTypes.string,
    number = _react.PropTypes.number,
    func = _react.PropTypes.func,
    oneOf = _react.PropTypes.oneOf;

var propTypes = {
  placeholder: string,
  searchBtnText: string,
  size: oneOf(['', 'lg']),
  onChange: func,
  delay: number
};
var defaultProps = {
  placeholder: 'Enter your search here...',
  searchBtnText: 'Search',
  size: '',
  delay: 500
};

var SearchInput = function (_PureComponent) {
  (0, _inherits3.default)(SearchInput, _PureComponent);

  function SearchInput(props) {
    (0, _classCallCheck3.default)(this, SearchInput);

    var _this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call(this, props));

    _this.debouncedOnChange = function (value) {
      var onChange = _this.props.onChange;


      if (typeof onChange === 'function') {
        onChange(value.trim());
      }
    };

    _this.handleChange = function (e) {
      var value = e.target.value;

      _this.setState({
        value: value
      }, function () {
        _this.debouncedOnChange(value);
      });
    };

    _this.handleCloseIconClick = function () {
      _this.setState({
        value: ''
      });
    };

    _this.state = {
      value: ''
    };
    _this.debouncedOnChange = (0, _utils.debounce)(_this.debouncedOnChange, props.delay);
    return _this;
  }

  SearchInput.prototype.render = function render() {
    /* eslint-disable no-unused-vars */
    var _props = this.props,
        placeholder = _props.placeholder,
        searchBtnText = _props.searchBtnText,
        size = _props.size,
        delay = _props.delay,
        rest = (0, _objectWithoutProperties3.default)(_props, ['placeholder', 'searchBtnText', 'size', 'delay']);
    var value = this.state.value;

    var hasFeedback = !!value;

    return _react2.default.createElement(
      'div',
      {
        className: (0, _utils.cls)(_templateObject, size ? 'search-box--' + size : '')
      },
      _react2.default.createElement(
        'div',
        {
          className: (0, _utils.cls)(_templateObject2, hasFeedback ? 'search-box__input-addon-group--has-feedback' : '')
        },
        _react2.default.createElement('i', { className: 'icon-search' }),
        _react2.default.createElement('input', (0, _extends3.default)({}, rest, {
          className: 'input',
          type: 'search',
          value: value,
          placeholder: placeholder,
          onChange: this.handleChange
        })),
        hasFeedback && _react2.default.createElement('span', { className: 'icon-close input-feedback active', onClick: this.handleCloseIconClick })
      ),
      _react2.default.createElement(
        'button',
        { type: 'button', className: 'btn btn-primary' },
        searchBtnText
      )
    );
  };

  return SearchInput;
}(_react.PureComponent);

exports.default = SearchInput;


SearchInput.displayName = 'SearchInput';
SearchInput.propTypes = propTypes;
SearchInput.defaultProps = defaultProps;
module.exports = exports['default'];
