// react-aaui/lib/TabsTitle.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _types = require('./shared/types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var string = _react.PropTypes.string,
    object = _react.PropTypes.object,
    any = _react.PropTypes.any;


var propTypes = {
  className: string,
  style: object,
  name: string,
  children: any,
  aauiTabsAPI: _types.tabsAPIShape
};

var TabsTitle = function TabsTitle(_ref) {
  var name = _ref.name,
      style = _ref.style,
      className = _ref.className,
      children = _ref.children,
      _ref$aauiTabsAPI = _ref.aauiTabsAPI,
      select = _ref$aauiTabsAPI.select,
      getSelected = _ref$aauiTabsAPI.getSelected;
  return _react2.default.createElement(
    'li',
    null,
    _react2.default.createElement(
      'a',
      {
        className: (name === getSelected() ? 'active' : '') + ' ' + (className || ''),
        style: style,
        onClick: function onClick() {
          return select(name);
        }
      },
      children
    )
  );
};

TabsTitle.displayName = 'AAUITabsTitle';
TabsTitle.propTypes = propTypes;

exports.default = TabsTitle;
module.exports = exports['default'];
