// react-aaui/lib/TabsContainer.js
'use strict';

exports.__esModule = true;

var _taggedTemplateLiteralLoose2 = require('babel-runtime/helpers/taggedTemplateLiteralLoose');

var _taggedTemplateLiteralLoose3 = _interopRequireDefault(_taggedTemplateLiteralLoose2);

var _templateObject = (0, _taggedTemplateLiteralLoose3.default)(['', ' ', ''], ['', ' ', '']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _utils = require('./shared/utils');

var _types = require('./shared/types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var string = _react.PropTypes.string,
    object = _react.PropTypes.object,
    any = _react.PropTypes.any;


var propTypes = {
  className: string,
  style: object,
  name: string,
  children: any,
  aauiTabsAPI: _types.tabsAPIShape
};

var TabsContainer = function TabsContainer(_ref) {
  var className = _ref.className,
      style = _ref.style,
      name = _ref.name,
      children = _ref.children,
      getSelected = _ref.aauiTabsAPI.getSelected;
  return _react2.default.createElement(
    'div',
    {
      className: (0, _utils.cls)(_templateObject, name !== getSelected() ? 'u-hidden' : '', className || ''),
      style: style
    },
    children
  );
};

TabsContainer.displayName = 'AAUITabsContainer';
TabsContainer.propTypes = propTypes;

exports.default = TabsContainer;
module.exports = exports['default'];
