// react-aaui/lib/Tag.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _taggedTemplateLiteralLoose2 = require('babel-runtime/helpers/taggedTemplateLiteralLoose');

var _taggedTemplateLiteralLoose3 = _interopRequireDefault(_taggedTemplateLiteralLoose2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _templateObject = (0, _taggedTemplateLiteralLoose3.default)(['aaui-tag\n                       aaui-tag-', '\n                       ', ''], ['aaui-tag\n                       aaui-tag-', '\n                       ', '']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _utils = require('./shared/utils');

require('./Tag.less');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = {
  className: _react.PropTypes.string,
  value: _react.PropTypes.string,
  onChange: _react.PropTypes.func,
  onCancel: _react.PropTypes.func,
  onClose: _react.PropTypes.func,
  editMode: _react.PropTypes.bool,
  isNew: _react.PropTypes.bool
};

var defaultProps = {
  editMode: false
};

var Tag = function (_PureComponent) {
  (0, _inherits3.default)(Tag, _PureComponent);

  function Tag(props) {
    (0, _classCallCheck3.default)(this, Tag);

    var _this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call(this, props));

    _this.convertEscEnterToBlur = function (e) {
      var target = e.target;
      switch (e.which) {
        case 27:
          e.preventDefault();
          _this.setState({
            isInputCancel: true,
            isTabEnter: false
          }, function () {
            target.blur();
          });
          break;
        case 9:
        case 13:
          e.preventDefault();
          _this.setState({
            isTabEnter: true
          }, function () {
            target.blur();
          });
          break;
        default:
          break;
      }
    };

    _this.changeValue = function (e) {
      var value = e.target.value;
      _this.setState({
        isInput: false
      }, function () {
        if (!_this.state.isInputCancel) {
          var onChange = _this.props.onChange;
          if (onChange) onChange(value, _this.state.isTabEnter);
        } else {
          var onCancel = _this.props.onCancel;
          if (onCancel) onCancel();
        }
      });
    };

    _this.toInput = function () {
      _this.setState({
        isInput: true,
        isInputCancel: false
      }, function () {
        var input = _this.input;
        input.focus();
        input.select();
      });
    };

    _this.toTag = function () {
      _this.setState({
        isInput: false
      });
    };

    _this.state = {
      isInput: false,
      isTabEnter: false
    };
    return _this;
  }

  Tag.prototype.render = function render() {
    var _this2 = this;

    /* eslint-disable no-unused-vars */
    var _props = this.props,
        onCancel = _props.onCancel,
        className = _props.className,
        editMode = _props.editMode,
        onClose = _props.onClose,
        onChange = _props.onChange,
        value = _props.value,
        isNew = _props.isNew,
        rest = (0, _objectWithoutProperties3.default)(_props, ['onCancel', 'className', 'editMode', 'onClose', 'onChange', 'value', 'isNew']);
    var isInput = this.state.isInput;


    return _react2.default.createElement(
      'span',
      (0, _extends3.default)({
        className: (0, _utils.cls)(_templateObject, editMode ? 'edit' : 'noedit', className || '')
      }, rest),
      isInput ? _react2.default.createElement('input', {
        ref: function ref(input) {
          _this2.input = input;
        },
        placeholder: isNew ? value : undefined,
        defaultValue: isNew ? undefined : value,
        onKeyDown: this.convertEscEnterToBlur,
        onBlur: this.changeValue
      }) : _react2.default.createElement(
        'span',
        { onClick: editMode ? this.toInput : undefined },
        value
      ),
      editMode && !isInput ? _react2.default.createElement('span', { className: 'aaui-tag-close aaui-icon-cross', onClick: onClose }) : undefined
    );
  };

  return Tag;
}(_react.PureComponent);

exports.default = Tag;


Tag.displayName = 'AAUITag';
Tag.propTypes = propTypes;
Tag.defaultProps = defaultProps;
module.exports = exports['default'];
