// react-aaui/lib/Tooltip.js
'use strict';

exports.__esModule = true;

var _taggedTemplateLiteralLoose2 = require('babel-runtime/helpers/taggedTemplateLiteralLoose');

var _taggedTemplateLiteralLoose3 = _interopRequireDefault(_taggedTemplateLiteralLoose2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _templateObject = (0, _taggedTemplateLiteralLoose3.default)(['tooltips\n                       ', '\n                       tooltips--', '\n                       ', ''], ['tooltips\n                       ', '\n                       tooltips--', '\n                       ', '']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('active.css/less/components/tooltip.less');

var _utils = require('./shared/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = {
  className: _react.PropTypes.string,
  style: _react.PropTypes.object,
  theme: _react.PropTypes.string,
  label: _react.PropTypes.string,
  direction: _react.PropTypes.string,
  children: _react.PropTypes.any
};

var defaultProps = {
  direction: 'n'
};

var Tooltip = function (_PureComponent) {
  (0, _inherits3.default)(Tooltip, _PureComponent);

  function Tooltip() {
    (0, _classCallCheck3.default)(this, Tooltip);
    return (0, _possibleConstructorReturn3.default)(this, _PureComponent.apply(this, arguments));
  }

  Tooltip.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        style = _props.style,
        theme = _props.theme,
        direction = _props.direction,
        children = _props.children,
        label = _props.label;


    return _react2.default.createElement(
      'span',
      {
        className: (0, _utils.cls)(_templateObject, theme ? 't-' + theme : '', direction, className || ''),
        style: style
      },
      label,
      _react2.default.createElement(
        'span',
        { className: 'tooltips__content' },
        children
      )
    );
  };

  return Tooltip;
}(_react.PureComponent);

exports.default = Tooltip;


Tooltip.displayName = 'AAUITooltip';
Tooltip.propTypes = propTypes;
Tooltip.defaultProps = defaultProps;
module.exports = exports['default'];
