// react-aaui/lib/shared/L10nProvider.js
'use strict';

exports.__esModule = true;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _react = require('react');

var _L10n = require('./L10n');

var _L10n2 = _interopRequireDefault(_L10n);

var _ValidationMsg = require('../form/i18n/ValidationMsg.json');

var _ValidationMsg2 = _interopRequireDefault(_ValidationMsg);

var _types = require('./types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var string = _react.PropTypes.string,
    object = _react.PropTypes.object,
    node = _react.PropTypes.node;

var propTypes = {
  locale: string,
  messages: object,
  children: node
};
var defaultProps = {
  locale: 'en_US'
};
var childContextTypes = (0, _extends3.default)({}, _types.aauiL10nPropTypes);

var L10nProvider = function (_PureComponent) {
  (0, _inherits3.default)(L10nProvider, _PureComponent);

  function L10nProvider(props, context) {
    (0, _classCallCheck3.default)(this, L10nProvider);

    var _this = (0, _possibleConstructorReturn3.default)(this, _PureComponent.call(this, props, context));

    _this.l10n = new _L10n2.default(props, context);
    return _this;
  }

  L10nProvider.prototype.getChildContext = function getChildContext() {
    return {
      aauiL10n: this.l10n
    };
  };

  L10nProvider.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    if (this.props.locale !== nextProps.locale || this.props.messages !== nextProps.messages) {
      this.l10n.locale = nextProps.locale;
      this.l10n.messages = (0, _extends3.default)({}, this.l10n.messages, _ValidationMsg2.default[nextProps.locale], nextProps.messages);
    }
  };

  L10nProvider.prototype.render = function render() {
    return _react.Children.only(this.props.children);
  };

  return L10nProvider;
}(_react.PureComponent);

exports.default = L10nProvider;


L10nProvider.propTypes = propTypes;
L10nProvider.defaultProps = defaultProps;
L10nProvider.childContextTypes = childContextTypes;
module.exports = exports['default'];
