// react-aaui/lib/shared/L10n.js
'use strict';

exports.__esModule = true;

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _createListenerCollection = require('./createListenerCollection');

var _createListenerCollection2 = _interopRequireDefault(_createListenerCollection);

var _ValidationMsg = require('../form/i18n/ValidationMsg.json');

var _ValidationMsg2 = _interopRequireDefault(_ValidationMsg);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var defaultValidationMsgKeys = {
  required: 'react-aaui.validation.required',
  limit: 'react-aaui.validation.limitation',
  max: 'react-aaui.validation.max',
  min: 'react-aaui.validation.min',
  invalid: 'react-aaui.validation.invalid'
};
var DEFAULT_LOCALE = 'en_US';

var L10n = function () {
  function L10n() {
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        _ref$locale = _ref.locale,
        locale = _ref$locale === undefined ? DEFAULT_LOCALE : _ref$locale,
        _ref$messages = _ref.messages,
        messages = _ref$messages === undefined ? {} : _ref$messages,
        _ref$config = _ref.config,
        config = _ref$config === undefined ? {} : _ref$config;

    (0, _classCallCheck3.default)(this, L10n);

    this._locale = locale;
    this._messages = (0, _extends3.default)({}, _ValidationMsg2.default[this._locale], messages);
    this._config = config;
    this.initDefaultMsg();

    this.listeners = (0, _createListenerCollection2.default)();
  }

  L10n.prototype.initDefaultMsg = function initDefaultMsg() {
    var _this = this;

    (0, _keys2.default)(defaultValidationMsgKeys).forEach(function (k) {
      _this[k] = defaultValidationMsgKeys[k];
    });
  };

  L10n.prototype.notifiy = function notifiy() {
    this.listeners.notify();
  };

  L10n.prototype.subscribe = function subscribe(listener) {
    if (typeof listener !== 'function') {
      throw new Error('Expected listener to be a function.');
    }

    return this.listeners.subscribe(listener);
  };

  (0, _createClass3.default)(L10n, [{
    key: 'locale',
    get: function get() {
      return this._locale;
    },
    set: function set(value) {
      this._locale = value;
      this.initDefaultMsg();
      this.notifiy();
    }
  }, {
    key: 'messages',
    get: function get() {
      return this._messages;
    },
    set: function set(value) {
      this._messages = value;
      this.initDefaultMsg();
      this.notifiy();
    }
  }]);
  return L10n;
}();

exports.default = L10n;
module.exports = exports['default'];
