// react-base-ui/src/confirmation/components/Dialog/index.jsx
import React, { PropTypes } from 'react';
import Button from 'react-aaui/lib/Button';
import Modal from 'react-aaui/lib/Modal';
import 'active.css/less/components/modal.less';

import './index.less';

export default class Dialog extends React.PureComponent {

  static propTypes = {
    show: PropTypes.bool,
    proceed: PropTypes.func,
    cancel: PropTypes.func,
    dismiss: PropTypes.func,
    message: PropTypes.oneOfType([PropTypes.array, PropTypes.element, PropTypes.string]),
    title: PropTypes.string,
    showCancel: PropTypes.bool,
    cancelText: PropTypes.string,
    confirmText: PropTypes.string
  }

  static defaultProps = {
    cancelText: 'Cancel',
    confirmText: 'OK'
  }

  render() {
    const {
      title = 'Confirm',
      message = '',
      show,
      dismiss,
      cancel,
      proceed,
      showCancel = false,
      cancelText = 'Cancel',
      confirmText = 'OK'
    } = this.props;

    return (
      <Modal shown={show} title={title} className="dialog" onClose={dismiss}>
        <div className="modal-body">
          { message }
        </div>
        <div className="modal-footer">
          {showCancel ?
            <Button type="secondary" onClick={() => cancel()}>
              {cancelText}
            </Button>
          : ''
          }

          <Button type="strong" onClick={() => proceed()}>
            {confirmText}
          </Button>
        </div>
      </Modal>
    );
  }
}
