// react-aaui/es/Button.js
import _extends from 'babel-runtime/helpers/extends';
import _taggedTemplateLiteralLoose from 'babel-runtime/helpers/taggedTemplateLiteralLoose';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _templateObject = _taggedTemplateLiteralLoose(['btn\n          btn-', '\n          ', '\n          ', ''], ['btn\n          btn-', '\n          ', '\n          ', '']);

import React, { PureComponent, PropTypes } from 'react';
import 'active.css/less/components/button.less';

import { cls } from './shared/utils';

var bool = PropTypes.bool,
    string = PropTypes.string,
    any = PropTypes.any;


var propTypes = {
  noSubmit: bool,
  type: string.isRequired,
  size: string,
  className: string,
  children: any
};

var defaultProps = {
  noSubmit: false,
  type: 'secondary'
};

var Button = function (_PureComponent) {
  _inherits(Button, _PureComponent);

  function Button() {
    _classCallCheck(this, Button);

    return _possibleConstructorReturn(this, _PureComponent.apply(this, arguments));
  }

  Button.prototype.render = function render() {
    var _props = this.props,
        noSubmit = _props.noSubmit,
        type = _props.type,
        size = _props.size,
        children = _props.children,
        className = _props.className,
        rest = _objectWithoutProperties(_props, ['noSubmit', 'type', 'size', 'children', 'className']);

    return React.createElement(
      'button',
      _extends({}, rest, {
        type: noSubmit ? 'button' : 'submit',
        className: cls(_templateObject, type, size ? 'btn--' + size : '', className || '')
      }),
      children
    );
  };

  return Button;
}(PureComponent);

export default Button;


Button.displayName = 'AAUIButton';
Button.propTypes = propTypes;
Button.defaultProps = defaultProps;
module.exports = exports['default'];
