// react-aaui/es/DatePicker.js
import _taggedTemplateLiteralLoose from 'babel-runtime/helpers/taggedTemplateLiteralLoose';
import _Object$defineProperties from 'babel-runtime/core-js/object/define-properties';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _templateObject = _taggedTemplateLiteralLoose(['date-picker ', ''], ['date-picker ', '']),
    _templateObject2 = _taggedTemplateLiteralLoose(['date-picker__calendar\n                         ', '\n                         ', '\n                         ', ''], ['date-picker__calendar\n                         ', '\n                         ', '\n                         ', '']),
    _templateObject3 = _taggedTemplateLiteralLoose(['', '\n      ', '\n      ', '\n      ', ''], ['', '\n      ', '\n      ', '\n      ', '']);

import React, { PureComponent, PropTypes } from 'react';
/* eslint import/no-extraneous-dependencies: 0 */
import ReactDOM from 'react-dom';
import 'active.css/less/components/date-picker.less';
import { cls } from './shared/utils';
import Input from './Input';

var MONTH_NAMES = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];

var DAY_SPAN = 1000 * 60 * 60 * 24;

function findOutDays(d) {
  var firstDayOfMonth = new Date(d.getFullYear(), d.getMonth(), 1);
  var startDate = new Date(firstDayOfMonth.getTime() - DAY_SPAN * firstDayOfMonth.getDay());

  function addDay(date, n) {
    var offset = date.getTimezoneOffset();
    var nextDay = new Date(date.getTime() + n * DAY_SPAN);
    var nextDayOffset = nextDay.getTimezoneOffset();
    if (nextDayOffset !== offset) {
      nextDay.setTime(nextDay.getTime() + (nextDayOffset - offset) * 60 * 1000);
    }
    return nextDay;
  }

  var days = new Array(6);
  for (var i = 0; i < 6; i += 1) {
    days[i] = new Array(7);
    for (var j = 0; j < 7; j += 1) {
      days[i][j] = addDay(startDate, i * 7 + j);
    }
  }
  return days;
}

function formatDate(d) {
  if (d == null) return '';
  return d.getMonth() + 1 + '/' + d.getDate() + '/' + d.getFullYear();
}

function isOtherMonth(x, y) {
  return x.getMonth() !== y.getMonth();
}

function isSameDay(x, y) {
  if (x === y) return true;
  if (x == null || y == null) return false;
  return x.getFullYear() === y.getFullYear() && x.getMonth() === y.getMonth() && x.getDate() === y.getDate();
}

var propTypes = {
  className: PropTypes.string,
  style: PropTypes.object,
  name: PropTypes.string,
  showIcon: PropTypes.bool,
  errored: PropTypes.bool,
  disabled: PropTypes.bool,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  transformDate: PropTypes.func,
  value: PropTypes.object,
  defaultValue: PropTypes.object,
  formatDate: PropTypes.func,
  today: PropTypes.object,
  getDateStatus: PropTypes.func,
  setDateClass: PropTypes.func,
  onClickDate: PropTypes.func,
  onChange: PropTypes.func
};

var DatePicker = function (_PureComponent) {
  _inherits(DatePicker, _PureComponent);

  function DatePicker(props) {
    _classCallCheck(this, DatePicker);

    var _this = _possibleConstructorReturn(this, _PureComponent.call(this, props));

    _this.componentDidMount = function () {
      _Object$defineProperties(_this, {
        value: {
          get: function get() {
            return this.state.value;
          },
          set: function set(v) {
            if (this.props.value === undefined) {
              this.setState({ value: v });
            }
          }
        }
      });
    };

    _this.componentWillReceiveProps = function (nextProps) {
      /* eslint no-nested-ternary: 0 */
      var newState = {
        value: nextProps.value !== _this.props.value ? nextProps.value : nextProps.defaultValue !== _this.props.defaultValue ? _this.state.value === undefined ? nextProps.defaultValue : _this.state.value : _this.state.value
      };
      if (newState.value !== _this.state.value) _this.setState(newState);
    };

    _this.componentDidUpdate = function () {
      var formatDateProps = _this.props.formatDate;
      var value = _this.state.value;


      if (formatDateProps) {
        _this.input.value = formatDateProps(value);
      } else if (value) {
        _this.input.value = formatDate(value);
      }
    };

    _this.onClickDate = function (d, dateStatusObj) {
      var onClickDate = _this.props.onClickDate;
      onClickDate = typeof onClickDate === 'function' ? onClickDate : null;

      if (onClickDate) {
        if (onClickDate(dateStatusObj)) {
          _this.selectDate(d);
        }
        return false;
      }

      _this.selectDate(d);
      return false;
    };

    _this.displayIfNotAlready = function () {
      if (document.activeElement === _this.input.refs.input && !_this.state.isDisplayed) {
        _this.setState({ isDisplayed: true });
      }
    };

    _this.preventStealingFocus = function (e) {
      e.preventDefault();
    };

    _this.goPrevMonth = function () {
      var calendarDate = _this.state.calendarDate;
      var prevMonth = new Date(new Date(calendarDate.getFullYear(), calendarDate.getMonth(), 1).getTime() - DAY_SPAN);
      _this.setState({
        calendarDate: prevMonth
      });
    };

    _this.goNextMonth = function () {
      var calendarDate = _this.state.calendarDate;
      var nextMonth = new Date(new Date(calendarDate.getFullYear(), calendarDate.getMonth(), 1).getTime() + DAY_SPAN * 31);
      _this.setState({
        calendarDate: nextMonth
      });
    };

    _this.selectDate = function (d) {
      var date = _this.props.value !== undefined ? _this.state.value : d;

      _this.setState({
        value: d && date,
        isDisplayed: false
      }, function () {
        if (_this.props.onChange) _this.props.onChange(d);
      });
    };

    _this.gainFocus = function () {
      /* eslint react/no-find-dom-node: 0 */
      var rect = ReactDOM.findDOMNode(_this.input).getBoundingClientRect();
      var verDistance = document.documentElement.clientHeight - (rect.top + rect.height);
      var horDistance = document.documentElement.clientWidth - (rect.left + rect.width);

      _this.setState({
        isDisplayed: true,
        showOnTop: verDistance < 216.75 && rect.top >= 216.75,
        showOnLeft: horDistance < 246 && rect.width <= 246,
        calendarDate: _this.state.value || new Date()
      });
    };

    _this.loseFocus = function () {
      var input = _this.input;
      var value = _this.state.value;

      var transformDate = _this.props.transformDate;
      var d = new Date(transformDate ? transformDate(input.value) : input.value);

      if (isNaN(d.getTime())) {
        _this.selectDate(null);
      } else if (value == null || d.getTime() !== value.getTime()) {
        _this.selectDate(d);
      } else {
        _this.setState({
          isDisplayed: false
        });
      }
    };

    _this.render = function () {
      var _this$props = _this.props,
          className = _this$props.className,
          style = _this$props.style,
          name = _this$props.name,
          showIcon = _this$props.showIcon,
          errored = _this$props.errored,
          disabled = _this$props.disabled,
          onFocus = _this$props.onFocus,
          onBlur = _this$props.onBlur;
      var _this$state = _this.state,
          isDisplayed = _this$state.isDisplayed,
          showOnTop = _this$state.showOnTop,
          showOnLeft = _this$state.showOnLeft;

      var calendarDate = _this.state.calendarDate;
      var calendarDateStr = MONTH_NAMES[calendarDate.getMonth()] + ' ' + calendarDate.getFullYear();
      var days = findOutDays(calendarDate);
      var today = _this.props.today || new Date();

      // set input element default value.
      var inputDefaultValue = _this.props.value ? _this.state.value : _this.props.defaultValue;
      inputDefaultValue = _this.props.formatDate ? _this.props.formatDate(inputDefaultValue) : formatDate(inputDefaultValue);
      var getDateStatus = _this.props.getDateStatus || null;

      return React.createElement(
        'div',
        {
          className: cls(_templateObject, className || ''),
          style: style
        },
        React.createElement(Input, {
          ref: function ref(_ref) {
            _this.input = _ref;
          },
          name: name,
          postText: showIcon ? '\uF073' : undefined,
          errored: errored,
          disabled: disabled,
          defaultValue: inputDefaultValue,
          onFocus: onFocus ? function (e) {
            _this.gainFocus();
            onFocus(e);
          } : _this.gainFocus,
          onBlur: onBlur ? function (e) {
            _this.loseFocus();
            onBlur(e);
          } : _this.loseFocus,
          onClick: _this.displayIfNotAlready
        }),
        React.createElement(
          'div',
          {
            className: cls(_templateObject2, !isDisplayed ? 'u-hidden' : '', showOnTop ? 'date-picker__calendar--show-on-top' : '', showOnLeft ? 'date-picker__calendar--show-on-left' : ''),
            onMouseDown: _this.preventStealingFocus
          },
          React.createElement(
            'header',
            { className: 'date-picker__header u-clearfix' },
            React.createElement('span', {
              className: 'icon-chevron-left date-picker__left-arrow',
              title: 'Previous month',
              onClick: _this.goPrevMonth
            }),
            React.createElement('span', {
              className: 'icon-chevron-right date-picker__right-arrow',
              title: 'Next month',
              onClick: _this.goNextMonth
            }),
            React.createElement(
              'h1',
              { className: 'date-picker__calendar-title' },
              calendarDateStr
            )
          ),
          React.createElement(
            'table',
            { className: 'date-picker__table' },
            React.createElement(
              'thead',
              null,
              React.createElement(
                'tr',
                null,
                React.createElement(
                  'th',
                  { title: 'Sunday' },
                  'S'
                ),
                React.createElement(
                  'th',
                  { title: 'Monday' },
                  'M'
                ),
                React.createElement(
                  'th',
                  { title: 'Tuesday' },
                  'T'
                ),
                React.createElement(
                  'th',
                  { title: 'Wednesday' },
                  'W'
                ),
                React.createElement(
                  'th',
                  { title: 'Thursday' },
                  'T'
                ),
                React.createElement(
                  'th',
                  { title: 'Friday' },
                  'F'
                ),
                React.createElement(
                  'th',
                  { title: 'Saturday' },
                  'S'
                )
              )
            ),
            React.createElement(
              'tbody',
              null,
              days.map(function (r, i) {
                return React.createElement(
                  'tr',
                  { key: i },
                  r.map(function (c, j) {
                    var dateStatusObj = getDateStatus && getDateStatus({
                      date: c,
                      calendarDate: calendarDate,
                      today: today
                    }) || null;

                    return React.createElement(
                      'td',
                      {
                        key: j,
                        className: _this.setDateClass(c, calendarDate, today, dateStatusObj),
                        onClick: function onClick() {
                          return _this.onClickDate(c, dateStatusObj);
                        }
                      },
                      c.getDate()
                    );
                  })
                );
              })
            )
          )
        )
      );
    };

    _this.state = {
      isDisplayed: false,
      showOnTop: false,
      showOnLeft: false,
      value: props.value || props.defaultValue,
      calendarDate: props.value || props.defaultValue || new Date()
    };
    return _this;
  }

  DatePicker.prototype.setDateClass = function setDateClass(c, calendarDate, today, dateStatusObj) {
    var setDateClass = this.props.setDateClass;
    setDateClass = typeof setDateClass === 'function' ? setDateClass : null;

    return cls(_templateObject3, isOtherMonth(c, calendarDate) ? 'date-picker--other-month' : '', isSameDay(c, today) ? 'date-picker--today' : '', isSameDay(c, this.state.value) ? 'date-picker--selected' : '', setDateClass && setDateClass(dateStatusObj));
  };

  return DatePicker;
}(PureComponent);

export default DatePicker;


DatePicker.displayName = 'AAUIDatePicker';
DatePicker.propTypes = propTypes;
module.exports = exports['default'];
