// react-aaui/es/ProgressSteps.js
import _taggedTemplateLiteralLoose from 'babel-runtime/helpers/taggedTemplateLiteralLoose';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _templateObject = _taggedTemplateLiteralLoose(['progress-steps ', ' ', ''], ['progress-steps ', ' ', '']);

import 'active.css/less/components/progress-steps.less';

import React, { PureComponent, PropTypes } from 'react';
import { cls } from './shared/utils';
import * as da from './shared/data-access';

var string = PropTypes.string,
    object = PropTypes.object,
    array = PropTypes.array;


var propTypes = {
  className: string,
  size: string,
  steps: array,
  style: object
};

var ProgressSteps = function (_PureComponent) {
  _inherits(ProgressSteps, _PureComponent);

  function ProgressSteps() {
    _classCallCheck(this, ProgressSteps);

    return _possibleConstructorReturn(this, _PureComponent.apply(this, arguments));
  }

  ProgressSteps.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        style = _props.style,
        size = _props.size,
        steps = _props.steps;

    return React.createElement(
      'ul',
      { className: cls(_templateObject, size ? 'progress-steps--' + size : '', className || ''), style: style },
      steps.map(function (step) {
        return React.createElement(
          'li',
          { key: da.get(step, 'text'), className: 'progress-steps__step ' + da.get(step, 'status') },
          React.createElement(
            'span',
            { className: 'progress-steps__step-text' },
            da.get(step, 'text')
          )
        );
      })
    );
  };

  return ProgressSteps;
}(PureComponent);

ProgressSteps.displayName = 'AAUIProgressSteps';
export default ProgressSteps;


ProgressSteps.displayName = 'AAUIProgressSteps';
ProgressSteps.propTypes = propTypes;
module.exports = exports['default'];
