// react-aaui/es/RadioGroup.js
import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { PureComponent, PropTypes } from 'react';

import 'active.css/less/components/radio-group.less';
import Radio from './Radio';

var string = PropTypes.string,
    func = PropTypes.func,
    node = PropTypes.node;

var propTypes = {
  children: node.isRequired,
  className: string,
  name: string,
  onChange: func,
  selected: string,
  size: string
};
var defaultProps = {
  className: 'radio-group'
};

var RadioGroup = function (_PureComponent) {
  _inherits(RadioGroup, _PureComponent);

  function RadioGroup() {
    _classCallCheck(this, RadioGroup);

    return _possibleConstructorReturn(this, _PureComponent.apply(this, arguments));
  }

  RadioGroup.prototype.render = function render() {
    var _props = this.props,
        name = _props.name,
        children = _props.children,
        className = _props.className,
        onChange = _props.onChange,
        selected = _props.selected,
        size = _props.size,
        rest = _objectWithoutProperties(_props, ['name', 'children', 'className', 'onChange', 'selected', 'size']);

    return React.createElement(
      'div',
      _extends({
        className: className
      }, rest),
      React.Children.map(children, function (child) {
        if (child.type === Radio) {
          return React.cloneElement(child, {
            name: name,
            onChange: onChange,
            checked: child.props.value === selected,
            size: size
          });
        }

        return child;
      })
    );
  };

  return RadioGroup;
}(PureComponent);

export default RadioGroup;


RadioGroup.displayName = 'RadioGroup';
RadioGroup.propTypes = propTypes;
RadioGroup.defaultProps = defaultProps;
module.exports = exports['default'];
