// react-aaui/es/Sidebar.js
import _taggedTemplateLiteralLoose from 'babel-runtime/helpers/taggedTemplateLiteralLoose';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';

var _templateObject = _taggedTemplateLiteralLoose(['sidebar\n          ', '\n          ', ''], ['sidebar\n          ', '\n          ', '']),
    _templateObject2 = _taggedTemplateLiteralLoose(['', '\n                      ', ''], ['', '\n                      ', '']);

import React, { PureComponent, PropTypes } from 'react';
import 'active.css/less/components/sidebar.less';

import { cls } from './shared/utils';
import * as da from './shared/data-access';

var bool = PropTypes.bool,
    string = PropTypes.string,
    number = PropTypes.number,
    object = PropTypes.object,
    array = PropTypes.array,
    element = PropTypes.element,
    oneOfType = PropTypes.oneOfType;


var propTypes = {
  className: string,
  style: object,
  data: array,
  links: bool,
  steps: bool,
  active: number,
  past: number,
  priority: string,
  title: string,
  linkElement: oneOfType([string, element])
};

var defaultProps = {
  priority: '5',
  steps: false,
  linkElement: 'a'
};

var Sidebar = function (_PureComponent) {
  _inherits(Sidebar, _PureComponent);

  function Sidebar() {
    _classCallCheck(this, Sidebar);

    return _possibleConstructorReturn(this, _PureComponent.apply(this, arguments));
  }

  Sidebar.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        style = _props.style,
        _props$data = _props.data,
        data = _props$data === undefined ? [] : _props$data,
        steps = _props.steps,
        active = _props.active,
        past = _props.past,
        priority = _props.priority,
        title = _props.title,
        links = _props.links,
        A = _props.linkElement;


    var HeaderElement = 'h' + priority;

    return React.createElement(
      'aside',
      {
        className: cls(_templateObject, links ? 'sidebar--links' : '', className || ''),
        style: style
      },
      React.createElement(
        HeaderElement,
        null,
        title
      ),
      React.createElement(
        'ul',
        null,
        data.map(function (item, i) {
          var stepNum = steps ? i + 1 + '.' : undefined;

          return React.createElement(
            'li',
            {
              key: i
            },
            React.createElement(
              A,
              {
                className: cls(_templateObject2, i === active ? 'active' : '', i === past ? 'past' : ''),
                href: da.get(item, 'href')
              },
              stepNum,
              ' ',
              da.get(item, 'text')
            )
          );
        })
      )
    );
  };

  return Sidebar;
}(PureComponent);

export default Sidebar;


Sidebar.displayName = 'AAUISidebar';
Sidebar.propTypes = propTypes;
Sidebar.defaultProps = defaultProps;
module.exports = exports['default'];
