// react-base-ui/lib/confirmation/confirmable.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var confirmable = function confirmable(Component) {
  var _class, _temp;

  return _temp = _class = function (_React$Component) {
    (0, _inherits3.default)(_class, _React$Component);

    function _class(props) {
      (0, _classCallCheck3.default)(this, _class);

      var _this = (0, _possibleConstructorReturn3.default)(this, _React$Component.call(this, props));

      _this.state = {
        show: true
      };
      return _this;
    }

    _class.prototype.dismiss = function dismiss() {
      var _this2 = this;

      this.setState({
        show: false
      }, function () {
        _this2.props.dispose();
      });
    };

    _class.prototype.cancel = function cancel(value) {
      var _this3 = this;

      this.setState({
        show: false
      }, function () {
        _this3.props.reject(value);
      });
    };

    _class.prototype.proceed = function proceed(value) {
      var _this4 = this;

      this.setState({
        show: false
      }, function () {
        _this4.props.resolve(value);
      });
    };

    _class.prototype.render = function render() {
      var _this5 = this;

      return _react2.default.createElement(Component, (0, _extends3.default)({
        proceed: function proceed() {
          return _this5.proceed();
        },
        cancel: function cancel() {
          return _this5.cancel();
        },
        dismiss: function dismiss() {
          return _this5.dismiss();
        },
        show: this.state.show }, this.props));
    };

    return _class;
  }(_react2.default.Component), _class.propTypes = {
    dispose: _react.PropTypes.func,
    reject: _react.PropTypes.func,
    resolve: _react.PropTypes.func
  }, _temp;
};

exports.default = confirmable;
module.exports = exports['default'];
