// react-base-ui/src/loading/Loading.jsx
import React from 'react';
import ReactDOM from 'react-dom';
import isElement from 'lodash/isElement';
import has from 'lodash/has';
import LoadingBar from './components/LoadingBar';

class Loading {
  constructor(targetElement = document.body, options = {}) {
    if (targetElement && !isElement(targetElement)) {
      options = targetElement;
      targetElement = document.body;
    }

    this.tartgetElement = targetElement;
    this.options = options;
    this.wrapperElement = this.tartgetElement.appendChild(document.createElement('div'));
    this.counter = 0;
    this.tranactions = {};
  }

  isLoading() {
    return this.counter > 0;
  }

  show() {
    if (this.counter === 0) {
      const { text = '' } = this.options;
      try {
        ReactDOM.render(
          <LoadingBar text={text} />,
        this.wrapperElement
      );
      } catch (e) {
        console.error(e);
        this.counter = -1;
      }
    }

    this.counter = this.counter + 1;
    return this;
  }

  hide() {
    if (this.counter > 0) {
      this.counter = this.counter - 1;

      if (this.counter === 0) {
        ReactDOM.unmountComponentAtNode(this.wrapperElement);
      }
    }
  }


  startTransaction(name = 'default') {
    if (has(this.tranactions, name)) {
      console.warn(`Transaction ${name} already exist`);
    } else {
      this.transactions[name] = setTimeout(() => {
        const id = this.transactions[name];
        if (id) {
          console.warn(`Transaction ${name} time out`);
          this.endTransaction(name);
        }
      }, 30000);
      this.show();
    }
  }

  endTransaction(name) {
    const id = this.transactions[name];
    if (id) {
      clearTimeout(id);
      delete this.tranactions[name];
      this.hide();
    } else {
      console.warn(`Transaction ${name} not found`);
    }
  }
}

export default Loading;
